{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating record set curation file\n",
    "\n",
    "- Fields in recordsets are annotated as a manual effort done by multiple people.\n",
    "- Curation was collected in a Google spreadsheet table.\n",
    "- The table has columns to make it easier for the curators to annotated also, also provides extra annotation for croissant ingestion\n",
    "\n",
    "## Columns\n",
    "\n",
    "- `dataset_name` - name of the dataset for curators.\n",
    "- `field_id` - identifier of the field for croissant.\n",
    "- `column_name` - just the label of the field for curators.\n",
    "- `column_description` - Curators annotate fields with description\n",
    "- `foreign_key` - curators add `field_id` of the foreign field\n",
    "- `bioregistry_prefix` - if data in a column comes from a database in bioregisty, annotate\n",
    "- `example` - helps curatos \n",
    "\n",
    "## Process\n",
    "\n",
    "1. Fetch curation table from Google\n",
    "2. Composing description\n",
    "3. Iterating over column and building output\n",
    "4. Save curation as json."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1256\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "# specifying the spreadsheet in the file 26.03:\n",
    "gid='328767591'\n",
    "\n",
    "# Curation of all columns from all OpenTargets output datasets:\n",
    "curation = f'https://docs.google.com/spreadsheets/d/132SKHMoaJePu4nTlBnQwfaz3dhfJiKmJUujfYkzXMdI/export?format=tsv&gid={gid}'\n",
    "# curation = 'annotated_new_evidence_columns.csv'\n",
    "# Folder to save the resulting curation file:\n",
    "asset_folder = '../src/ot_croissant/assets/'\n",
    "\n",
    "# Reading table:\n",
    "curation_table = (\n",
    "    pd.read_csv(curation, sep='\\t')\n",
    "    .astype(\n",
    "        {\n",
    "            'column_description': pd.StringDtype(),\n",
    "            'foreign_key': pd.StringDtype(),\n",
    "        }\n",
    "    )\n",
    "    .sort_values(['dataset_name', 'field_id'])\n",
    ")\n",
    "curation_table.head()\n",
    "print(len(curation_table))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collection of curated dataset:\n",
    "curation_json = []\n",
    "\n",
    "# Composing description:\n",
    "def compose_description(row: pd.Series) -> str:\n",
    "    \"\"\"\n",
    "    Composes the description of a column based on the bioregistry prefix and the column description.\n",
    "    If the bioregistry prefix is not available, it returns the column description as is.\n",
    "\n",
    "    Args:\n",
    "        row (pd.Series): A row from the curation table.\n",
    "    \n",
    "    Returns:\n",
    "        str: The composed description.\n",
    "    \"\"\"\n",
    "    # If the bioregistry prefix is not available, return the column description as is:\n",
    "    description = (\n",
    "        row['column_description']\n",
    "        if pd.isna(row['bioregistry_prefix'])\n",
    "        else f\"{row['column_description']} [bioregistry:{row['bioregistry_prefix'].lower()}]\"\n",
    "    )\n",
    "\n",
    "    return description\n",
    "\n",
    "# Iterating over the rows of the curation table:\n",
    "for _, row in curation_table.iterrows():\n",
    "    # If the column description is not available, skip the row:\n",
    "    if pd.isna(row['column_description']):\n",
    "        continue\n",
    "\n",
    "    # Adding curation to the dictionary:\n",
    "    data = {\n",
    "        'id': row['field_id'],\n",
    "        'description': compose_description(row)\n",
    "    }\n",
    "\n",
    "    # If the foreign key is available, add it to the dictionary:\n",
    "    if not pd.isna(row['foreign_key']):\n",
    "        data['foreign_key'] = row['foreign_key']\n",
    "\n",
    "    # If the bioregistry prefix is available, add it to the dictionary:\n",
    "    curation_json.append(data)\n",
    "\n",
    "# Saving the curation to a JSON file:\n",
    "with open(f'{asset_folder}/recordset.json', 'w') as f:\n",
    "    json.dump(curation_json, f, indent=2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>field_id</th>\n",
       "      <th>column_name</th>\n",
       "      <th>column_description</th>\n",
       "      <th>foreign_key</th>\n",
       "      <th>bioregistry_prefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>biosample</td>\n",
       "      <td>biosample/biosampleId</td>\n",
       "      <td>biosampleId</td>\n",
       "      <td>Unique identifier for the biosample</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>UBERON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disease_phenotype</td>\n",
       "      <td>disease_phenotype/evidence/references</td>\n",
       "      <td>references</td>\n",
       "      <td>References or citations supporting the evidence</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>evidence_cancer_biomarkers</td>\n",
       "      <td>evidence_cancer_biomarkers/drugId</td>\n",
       "      <td>drugId</td>\n",
       "      <td>Open Targets drug identifier</td>\n",
       "      <td>drug_molecule/id</td>\n",
       "      <td>CHEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>evidence_chembl</td>\n",
       "      <td>evidence_chembl/drugId</td>\n",
       "      <td>drugId</td>\n",
       "      <td>Open Targets drug identifier</td>\n",
       "      <td>drug_molecule/id</td>\n",
       "      <td>CHEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388</th>\n",
       "      <td>evidence_eva</td>\n",
       "      <td>evidence_eva/variantFunctionalConsequenceId</td>\n",
       "      <td>variantFunctionalConsequenceId</td>\n",
       "      <td>Sequence ontology (SO) identifier of the funct...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>evidence_eva_somatic</td>\n",
       "      <td>evidence_eva_somatic/variantFunctionalConseque...</td>\n",
       "      <td>variantFunctionalConsequenceId</td>\n",
       "      <td>Sequence ontology (SO) identifier of the funct...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>evidence_gene2phenotype</td>\n",
       "      <td>evidence_gene2phenotype/variantFunctionalConse...</td>\n",
       "      <td>variantFunctionalConsequenceId</td>\n",
       "      <td>Sequence ontology (SO) identifier of the funct...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>evidence_gene_burden</td>\n",
       "      <td>evidence_gene_burden/ancestryId</td>\n",
       "      <td>ancestryId</td>\n",
       "      <td>Identifier of the ancestry in the HANCESTRO on...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>HANCESTRO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>554</th>\n",
       "      <td>evidence_impc</td>\n",
       "      <td>evidence_impc/targetInModelEnsemblId</td>\n",
       "      <td>targetInModelEnsemblId</td>\n",
       "      <td>Target Ensembl ID in animal model</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>evidence_impc</td>\n",
       "      <td>evidence_impc/targetInModelMgiId</td>\n",
       "      <td>targetInModelMgiId</td>\n",
       "      <td>Target MGI ID in animal model</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>evidence_orphanet</td>\n",
       "      <td>evidence_orphanet/variantFunctionalConsequenceId</td>\n",
       "      <td>variantFunctionalConsequenceId</td>\n",
       "      <td>Sequence ontology (SO) identifier of the funct...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>695</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Ensembl human gene identifier for the expresse...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/tissues/efo_code</td>\n",
       "      <td>efo_code</td>\n",
       "      <td>Ontology ID of the biosample the expression da...</td>\n",
       "      <td>biosample/biosampleId</td>\n",
       "      <td>UBERON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713</th>\n",
       "      <td>go</td>\n",
       "      <td>go/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Gene ontology term identifier</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>go</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>730</th>\n",
       "      <td>interaction</td>\n",
       "      <td>interaction/targetA</td>\n",
       "      <td>targetA</td>\n",
       "      <td>Open Targets target identifier of the first mo...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731</th>\n",
       "      <td>interaction</td>\n",
       "      <td>interaction/targetB</td>\n",
       "      <td>targetB</td>\n",
       "      <td>Open Targets target identifier of the second m...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/expansionMethodMiIdentifier</td>\n",
       "      <td>expansionMethodMiIdentifier</td>\n",
       "      <td>Molecular Interactions (MI) identifier for the...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/interactionDetectionMetho...</td>\n",
       "      <td>interactionDetectionMethodMiIdentifier</td>\n",
       "      <td>Molecular Interactions (MI) identifier for the...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>754</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/interactionTypeMiIdentifier</td>\n",
       "      <td>interactionTypeMiIdentifier</td>\n",
       "      <td>Molecular Interactions (MI) identifier for the...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/participantDetectionMetho...</td>\n",
       "      <td>miIdentifier</td>\n",
       "      <td>Molecular Interactions (MI) identifier for par...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>760</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/participantDetectionMetho...</td>\n",
       "      <td>miIdentifier</td>\n",
       "      <td>Molecular Interactions (MI) identifier for par...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/targetA</td>\n",
       "      <td>targetA</td>\n",
       "      <td>Open Targets target identifier of the second m...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>772</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/targetB</td>\n",
       "      <td>targetB</td>\n",
       "      <td>Open Targets target identifier of the second m...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779</th>\n",
       "      <td>interval</td>\n",
       "      <td>interval/geneId</td>\n",
       "      <td>geneId</td>\n",
       "      <td>Ensembl gene identifier of the target the inte...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>832</th>\n",
       "      <td>mouse_phenotype</td>\n",
       "      <td>mouse_phenotype/biologicalModels/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Unique identifier for the biological model</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>833</th>\n",
       "      <td>mouse_phenotype</td>\n",
       "      <td>mouse_phenotype/biologicalModels/literature</td>\n",
       "      <td>literature</td>\n",
       "      <td>References related to the mouse model</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835</th>\n",
       "      <td>mouse_phenotype</td>\n",
       "      <td>mouse_phenotype/modelPhenotypeClasses/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Unique identifier for the phenotype class</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>837</th>\n",
       "      <td>mouse_phenotype</td>\n",
       "      <td>mouse_phenotype/modelPhenotypeId</td>\n",
       "      <td>modelPhenotypeId</td>\n",
       "      <td>Identifier for the specific phenotype observed...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>841</th>\n",
       "      <td>mouse_phenotype</td>\n",
       "      <td>mouse_phenotype/targetInModelEnsemblId</td>\n",
       "      <td>targetInModelEnsemblId</td>\n",
       "      <td>Ensembl identifier for the target gene in the ...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>842</th>\n",
       "      <td>mouse_phenotype</td>\n",
       "      <td>mouse_phenotype/targetInModelMgiId</td>\n",
       "      <td>targetInModelMgiId</td>\n",
       "      <td>Mouse Genome Informatics (MGI) identifier for ...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>868</th>\n",
       "      <td>pharmacogenomics</td>\n",
       "      <td>pharmacogenomics/drugs/drugId</td>\n",
       "      <td>drugId</td>\n",
       "      <td>List of Open Targets drug identifiers</td>\n",
       "      <td>drug_molecule/id</td>\n",
       "      <td>chembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>891</th>\n",
       "      <td>pharmacogenomics</td>\n",
       "      <td>pharmacogenomics/variantFunctionalConsequenceId</td>\n",
       "      <td>variantFunctionalConsequenceId</td>\n",
       "      <td>The sequence ontology identifier of the conseq...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>897</th>\n",
       "      <td>reactome</td>\n",
       "      <td>reactome/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Unique identifier for the Reactome pathway</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>reactome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901</th>\n",
       "      <td>so</td>\n",
       "      <td>so/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Sequence ontology term identifier</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>so</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907</th>\n",
       "      <td>study</td>\n",
       "      <td>study/biosampleId</td>\n",
       "      <td>biosampleId</td>\n",
       "      <td>Biosample identifier that maps to the biosampl...</td>\n",
       "      <td>biosample/biosampleId</td>\n",
       "      <td>UBERON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914</th>\n",
       "      <td>study</td>\n",
       "      <td>study/geneId</td>\n",
       "      <td>geneId</td>\n",
       "      <td>Ensembl identifier of the gene representing a ...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>928</th>\n",
       "      <td>study</td>\n",
       "      <td>study/pubmedId</td>\n",
       "      <td>pubmedId</td>\n",
       "      <td>PubMed identifier of the publication that refe...</td>\n",
       "      <td>literature/pmid</td>\n",
       "      <td>PUBMED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>target</td>\n",
       "      <td>target/go/ecoId</td>\n",
       "      <td>ecoId</td>\n",
       "      <td>Evidence &amp; Conclusion Ontology (ECO) identifie...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ECO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>target</td>\n",
       "      <td>target/go/geneProduct</td>\n",
       "      <td>geneProduct</td>\n",
       "      <td>Gene product associated with the GO annotation</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>uniprot</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>target</td>\n",
       "      <td>target/go/id</td>\n",
       "      <td>id</td>\n",
       "      <td>GO identifier for the annotation</td>\n",
       "      <td>go/id</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>target</td>\n",
       "      <td>target/hallmarks/attributes/pmid</td>\n",
       "      <td>pmid</td>\n",
       "      <td>PubMed ID of the supporting literature for the...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1003</th>\n",
       "      <td>target</td>\n",
       "      <td>target/hallmarks/cancerHallmarks/pmid</td>\n",
       "      <td>pmid</td>\n",
       "      <td>PubMed ID of the supporting literature for the...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1014</th>\n",
       "      <td>target</td>\n",
       "      <td>target/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Unique identifier for the target</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053</th>\n",
       "      <td>target</td>\n",
       "      <td>target/subcellularLocations/termSL</td>\n",
       "      <td>termSL</td>\n",
       "      <td>Subcellular location term identifier from Swis...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>SL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1085</th>\n",
       "      <td>target_essentiality</td>\n",
       "      <td>target_essentiality/geneEssentiality/depMapEss...</td>\n",
       "      <td>tissueId</td>\n",
       "      <td>Identifier of the tissue from which the cells ...</td>\n",
       "      <td>biosample/biosampleId</td>\n",
       "      <td>UBERON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1088</th>\n",
       "      <td>target_essentiality</td>\n",
       "      <td>target_essentiality/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Target identifier the essentiality was measure...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1115</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/mostSevereConsequenceId</td>\n",
       "      <td>mostSevereConsequenceId</td>\n",
       "      <td>Sequence ontology identifier of the most sever...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1134</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/transcriptConsequences/targetId</td>\n",
       "      <td>targetId</td>\n",
       "      <td>Open Target target identifier of the transcript</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1135</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/transcriptConsequences/transcriptId</td>\n",
       "      <td>transcriptId</td>\n",
       "      <td>Ensembl transcript identifier</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1137</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/transcriptConsequences/uniprotAccessions</td>\n",
       "      <td>uniprotAccessions</td>\n",
       "      <td>Uniprot identifiers of the gene product</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>UNIPROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1138</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/transcriptConsequences/variantFunction...</td>\n",
       "      <td>variantFunctionalConsequenceIds</td>\n",
       "      <td>The sequence ontology identifier of the conseq...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1146</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/variantEffect/targetId</td>\n",
       "      <td>targetId</td>\n",
       "      <td>Target identifier on which the variant effect ...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    dataset_name  \\\n",
       "31                     biosample   \n",
       "140            disease_phenotype   \n",
       "213   evidence_cancer_biomarkers   \n",
       "252              evidence_chembl   \n",
       "388                 evidence_eva   \n",
       "416         evidence_eva_somatic   \n",
       "459      evidence_gene2phenotype   \n",
       "462         evidence_gene_burden   \n",
       "554                evidence_impc   \n",
       "555                evidence_impc   \n",
       "596            evidence_orphanet   \n",
       "695                   expression   \n",
       "698                   expression   \n",
       "713                           go   \n",
       "730                  interaction   \n",
       "731                  interaction   \n",
       "733         interaction_evidence   \n",
       "747         interaction_evidence   \n",
       "754         interaction_evidence   \n",
       "757         interaction_evidence   \n",
       "760         interaction_evidence   \n",
       "771         interaction_evidence   \n",
       "772         interaction_evidence   \n",
       "779                     interval   \n",
       "832              mouse_phenotype   \n",
       "833              mouse_phenotype   \n",
       "835              mouse_phenotype   \n",
       "837              mouse_phenotype   \n",
       "841              mouse_phenotype   \n",
       "842              mouse_phenotype   \n",
       "868             pharmacogenomics   \n",
       "891             pharmacogenomics   \n",
       "897                     reactome   \n",
       "901                           so   \n",
       "907                        study   \n",
       "914                        study   \n",
       "928                        study   \n",
       "989                       target   \n",
       "991                       target   \n",
       "992                       target   \n",
       "998                       target   \n",
       "1003                      target   \n",
       "1014                      target   \n",
       "1053                      target   \n",
       "1085         target_essentiality   \n",
       "1088         target_essentiality   \n",
       "1115                     variant   \n",
       "1134                     variant   \n",
       "1135                     variant   \n",
       "1137                     variant   \n",
       "1138                     variant   \n",
       "1146                     variant   \n",
       "\n",
       "                                               field_id  \\\n",
       "31                                biosample/biosampleId   \n",
       "140               disease_phenotype/evidence/references   \n",
       "213                   evidence_cancer_biomarkers/drugId   \n",
       "252                              evidence_chembl/drugId   \n",
       "388         evidence_eva/variantFunctionalConsequenceId   \n",
       "416   evidence_eva_somatic/variantFunctionalConseque...   \n",
       "459   evidence_gene2phenotype/variantFunctionalConse...   \n",
       "462                     evidence_gene_burden/ancestryId   \n",
       "554                evidence_impc/targetInModelEnsemblId   \n",
       "555                    evidence_impc/targetInModelMgiId   \n",
       "596    evidence_orphanet/variantFunctionalConsequenceId   \n",
       "695                                       expression/id   \n",
       "698                         expression/tissues/efo_code   \n",
       "713                                               go/id   \n",
       "730                                 interaction/targetA   \n",
       "731                                 interaction/targetB   \n",
       "733    interaction_evidence/expansionMethodMiIdentifier   \n",
       "747   interaction_evidence/interactionDetectionMetho...   \n",
       "754    interaction_evidence/interactionTypeMiIdentifier   \n",
       "757   interaction_evidence/participantDetectionMetho...   \n",
       "760   interaction_evidence/participantDetectionMetho...   \n",
       "771                        interaction_evidence/targetA   \n",
       "772                        interaction_evidence/targetB   \n",
       "779                                     interval/geneId   \n",
       "832                 mouse_phenotype/biologicalModels/id   \n",
       "833         mouse_phenotype/biologicalModels/literature   \n",
       "835            mouse_phenotype/modelPhenotypeClasses/id   \n",
       "837                    mouse_phenotype/modelPhenotypeId   \n",
       "841              mouse_phenotype/targetInModelEnsemblId   \n",
       "842                  mouse_phenotype/targetInModelMgiId   \n",
       "868                       pharmacogenomics/drugs/drugId   \n",
       "891     pharmacogenomics/variantFunctionalConsequenceId   \n",
       "897                                         reactome/id   \n",
       "901                                               so/id   \n",
       "907                                   study/biosampleId   \n",
       "914                                        study/geneId   \n",
       "928                                      study/pubmedId   \n",
       "989                                     target/go/ecoId   \n",
       "991                               target/go/geneProduct   \n",
       "992                                        target/go/id   \n",
       "998                    target/hallmarks/attributes/pmid   \n",
       "1003              target/hallmarks/cancerHallmarks/pmid   \n",
       "1014                                          target/id   \n",
       "1053                 target/subcellularLocations/termSL   \n",
       "1085  target_essentiality/geneEssentiality/depMapEss...   \n",
       "1088                             target_essentiality/id   \n",
       "1115                    variant/mostSevereConsequenceId   \n",
       "1134            variant/transcriptConsequences/targetId   \n",
       "1135        variant/transcriptConsequences/transcriptId   \n",
       "1137   variant/transcriptConsequences/uniprotAccessions   \n",
       "1138  variant/transcriptConsequences/variantFunction...   \n",
       "1146                     variant/variantEffect/targetId   \n",
       "\n",
       "                                 column_name  \\\n",
       "31                               biosampleId   \n",
       "140                               references   \n",
       "213                                   drugId   \n",
       "252                                   drugId   \n",
       "388           variantFunctionalConsequenceId   \n",
       "416           variantFunctionalConsequenceId   \n",
       "459           variantFunctionalConsequenceId   \n",
       "462                               ancestryId   \n",
       "554                   targetInModelEnsemblId   \n",
       "555                       targetInModelMgiId   \n",
       "596           variantFunctionalConsequenceId   \n",
       "695                                       id   \n",
       "698                                 efo_code   \n",
       "713                                       id   \n",
       "730                                  targetA   \n",
       "731                                  targetB   \n",
       "733              expansionMethodMiIdentifier   \n",
       "747   interactionDetectionMethodMiIdentifier   \n",
       "754              interactionTypeMiIdentifier   \n",
       "757                             miIdentifier   \n",
       "760                             miIdentifier   \n",
       "771                                  targetA   \n",
       "772                                  targetB   \n",
       "779                                   geneId   \n",
       "832                                       id   \n",
       "833                               literature   \n",
       "835                                       id   \n",
       "837                         modelPhenotypeId   \n",
       "841                   targetInModelEnsemblId   \n",
       "842                       targetInModelMgiId   \n",
       "868                                   drugId   \n",
       "891           variantFunctionalConsequenceId   \n",
       "897                                       id   \n",
       "901                                       id   \n",
       "907                              biosampleId   \n",
       "914                                   geneId   \n",
       "928                                 pubmedId   \n",
       "989                                    ecoId   \n",
       "991                              geneProduct   \n",
       "992                                       id   \n",
       "998                                     pmid   \n",
       "1003                                    pmid   \n",
       "1014                                      id   \n",
       "1053                                  termSL   \n",
       "1085                                tissueId   \n",
       "1088                                      id   \n",
       "1115                 mostSevereConsequenceId   \n",
       "1134                                targetId   \n",
       "1135                            transcriptId   \n",
       "1137                       uniprotAccessions   \n",
       "1138         variantFunctionalConsequenceIds   \n",
       "1146                                targetId   \n",
       "\n",
       "                                     column_description  \\\n",
       "31                  Unique identifier for the biosample   \n",
       "140     References or citations supporting the evidence   \n",
       "213                        Open Targets drug identifier   \n",
       "252                        Open Targets drug identifier   \n",
       "388   Sequence ontology (SO) identifier of the funct...   \n",
       "416   Sequence ontology (SO) identifier of the funct...   \n",
       "459   Sequence ontology (SO) identifier of the funct...   \n",
       "462   Identifier of the ancestry in the HANCESTRO on...   \n",
       "554                   Target Ensembl ID in animal model   \n",
       "555                       Target MGI ID in animal model   \n",
       "596   Sequence ontology (SO) identifier of the funct...   \n",
       "695   Ensembl human gene identifier for the expresse...   \n",
       "698   Ontology ID of the biosample the expression da...   \n",
       "713                       Gene ontology term identifier   \n",
       "730   Open Targets target identifier of the first mo...   \n",
       "731   Open Targets target identifier of the second m...   \n",
       "733   Molecular Interactions (MI) identifier for the...   \n",
       "747   Molecular Interactions (MI) identifier for the...   \n",
       "754   Molecular Interactions (MI) identifier for the...   \n",
       "757   Molecular Interactions (MI) identifier for par...   \n",
       "760   Molecular Interactions (MI) identifier for par...   \n",
       "771   Open Targets target identifier of the second m...   \n",
       "772   Open Targets target identifier of the second m...   \n",
       "779   Ensembl gene identifier of the target the inte...   \n",
       "832          Unique identifier for the biological model   \n",
       "833               References related to the mouse model   \n",
       "835           Unique identifier for the phenotype class   \n",
       "837   Identifier for the specific phenotype observed...   \n",
       "841   Ensembl identifier for the target gene in the ...   \n",
       "842   Mouse Genome Informatics (MGI) identifier for ...   \n",
       "868               List of Open Targets drug identifiers   \n",
       "891   The sequence ontology identifier of the conseq...   \n",
       "897          Unique identifier for the Reactome pathway   \n",
       "901                   Sequence ontology term identifier   \n",
       "907   Biosample identifier that maps to the biosampl...   \n",
       "914   Ensembl identifier of the gene representing a ...   \n",
       "928   PubMed identifier of the publication that refe...   \n",
       "989   Evidence & Conclusion Ontology (ECO) identifie...   \n",
       "991      Gene product associated with the GO annotation   \n",
       "992                    GO identifier for the annotation   \n",
       "998   PubMed ID of the supporting literature for the...   \n",
       "1003  PubMed ID of the supporting literature for the...   \n",
       "1014                   Unique identifier for the target   \n",
       "1053  Subcellular location term identifier from Swis...   \n",
       "1085  Identifier of the tissue from which the cells ...   \n",
       "1088  Target identifier the essentiality was measure...   \n",
       "1115  Sequence ontology identifier of the most sever...   \n",
       "1134    Open Target target identifier of the transcript   \n",
       "1135                      Ensembl transcript identifier   \n",
       "1137            Uniprot identifiers of the gene product   \n",
       "1138  The sequence ontology identifier of the conseq...   \n",
       "1146  Target identifier on which the variant effect ...   \n",
       "\n",
       "                foreign_key bioregistry_prefix  \n",
       "31                     <NA>             UBERON  \n",
       "140                    <NA>             pubmed  \n",
       "213        drug_molecule/id             CHEMBL  \n",
       "252        drug_molecule/id             CHEMBL  \n",
       "388                   so/id                 SO  \n",
       "416                   so/id                 SO  \n",
       "459                   so/id                 SO  \n",
       "462                    <NA>          HANCESTRO  \n",
       "554                    <NA>            ENSEMBL  \n",
       "555                    <NA>                MGI  \n",
       "596                   so/id                 SO  \n",
       "695                    <NA>            ENSEMBL  \n",
       "698   biosample/biosampleId             UBERON  \n",
       "713                    <NA>                 go  \n",
       "730               target/id            ENSEMBL  \n",
       "731               target/id            ENSEMBL  \n",
       "733                    <NA>                 MI  \n",
       "747                    <NA>                 MI  \n",
       "754                    <NA>                 MI  \n",
       "757                    <NA>                 MI  \n",
       "760                    <NA>                 MI  \n",
       "771               target/id            ENSEMBL  \n",
       "772               target/id            ENSEMBL  \n",
       "779               target/id            ENSEMBL  \n",
       "832                    <NA>                MGI  \n",
       "833                    <NA>             pubmed  \n",
       "835                    <NA>                 MP  \n",
       "837                    <NA>                 MP  \n",
       "841                    <NA>            ENSEMBL  \n",
       "842                    <NA>                MGI  \n",
       "868        drug_molecule/id             chembl  \n",
       "891                   so/id                 SO  \n",
       "897                    <NA>           reactome  \n",
       "901                    <NA>                 so  \n",
       "907   biosample/biosampleId             UBERON  \n",
       "914               target/id            ENSEMBL  \n",
       "928         literature/pmid             PUBMED  \n",
       "989                    <NA>                ECO  \n",
       "991                    <NA>            uniprot  \n",
       "992                   go/id                 GO  \n",
       "998                    <NA>             pubmed  \n",
       "1003                   <NA>             pubmed  \n",
       "1014                   <NA>            ENSEMBL  \n",
       "1053                   <NA>                 SL  \n",
       "1085  biosample/biosampleId             UBERON  \n",
       "1088              target/id            ENSEMBL  \n",
       "1115                  so/id                 SO  \n",
       "1134              target/id            ENSEMBL  \n",
       "1135                   <NA>            ENSEMBL  \n",
       "1137                   <NA>            UNIPROT  \n",
       "1138                  so/id                 SO  \n",
       "1146              target/id            ENSEMBL  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Columns with bioregistry prefix:\n",
    "curation_table.loc[curation_table.bioregistry_prefix.notna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>field_id</th>\n",
       "      <th>column_name</th>\n",
       "      <th>column_description</th>\n",
       "      <th>foreign_key</th>\n",
       "      <th>bioregistry_prefix</th>\n",
       "      <th>Example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/tissues/protein/reliability</td>\n",
       "      <td>reliability</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/tissues/protein/cell_type/reliability</td>\n",
       "      <td>reliability</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/tissues/protein/cell_type/level</td>\n",
       "      <td>level</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   dataset_name                                          field_id  \\\n",
       "67   expression            expression/tissues/protein/reliability   \n",
       "71   expression  expression/tissues/protein/cell_type/reliability   \n",
       "72   expression        expression/tissues/protein/cell_type/level   \n",
       "\n",
       "    column_name column_description foreign_key bioregistry_prefix Example  \n",
       "67  reliability               <NA>        <NA>                NaN     NaN  \n",
       "71  reliability               <NA>        <NA>                NaN     NaN  \n",
       "72        level               <NA>        <NA>                NaN     NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fields without description:\n",
    "curation_table.loc[curation_table.column_description.isna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Adding tags to distributions\n",
    "\n",
    "\n",
    "Creat distribution json based on curation table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>nice_name</th>\n",
       "      <th>key</th>\n",
       "      <th>description</th>\n",
       "      <th>tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>association_by_datasource_direct</td>\n",
       "      <td>Associations - direct (by data source)</td>\n",
       "      <td>[association_by_datasource_direct/diseaseId, a...</td>\n",
       "      <td>Data source-specific metrics for direct target...</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>association_by_datasource_indirect</td>\n",
       "      <td>Associations - indirect (by data source)</td>\n",
       "      <td>[association_by_datasource_indirect/diseaseId,...</td>\n",
       "      <td>Data source-specific metrics for direct target...</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>association_by_datatype_direct</td>\n",
       "      <td>Associations - direct (by data type)</td>\n",
       "      <td>[association_by_datatype_direct/diseaseId, ass...</td>\n",
       "      <td>Data type-specific metrics for direct target-d...</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>association_by_datatype_indirect</td>\n",
       "      <td>Associations - indirect (by data type)</td>\n",
       "      <td>[association_by_datatype_indirect/diseaseId, a...</td>\n",
       "      <td>Data type-specific metrics for direct target-d...</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>association_overall_direct</td>\n",
       "      <td>Associations - direct (overall score)</td>\n",
       "      <td>[association_overall_direct/diseaseId, associa...</td>\n",
       "      <td>Overall metrics for direct target-disease asso...</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   id  \\\n",
       "0    association_by_datasource_direct   \n",
       "1  association_by_datasource_indirect   \n",
       "2      association_by_datatype_direct   \n",
       "3    association_by_datatype_indirect   \n",
       "4          association_overall_direct   \n",
       "\n",
       "                                  nice_name  \\\n",
       "0    Associations - direct (by data source)   \n",
       "1  Associations - indirect (by data source)   \n",
       "2      Associations - direct (by data type)   \n",
       "3    Associations - indirect (by data type)   \n",
       "4     Associations - direct (overall score)   \n",
       "\n",
       "                                                 key  \\\n",
       "0  [association_by_datasource_direct/diseaseId, a...   \n",
       "1  [association_by_datasource_indirect/diseaseId,...   \n",
       "2  [association_by_datatype_direct/diseaseId, ass...   \n",
       "3  [association_by_datatype_indirect/diseaseId, a...   \n",
       "4  [association_overall_direct/diseaseId, associa...   \n",
       "\n",
       "                                         description              tags  \n",
       "0  Data source-specific metrics for direct target...  [Target-Disease]  \n",
       "1  Data source-specific metrics for direct target...  [Target-Disease]  \n",
       "2  Data type-specific metrics for direct target-d...  [Target-Disease]  \n",
       "3  Data type-specific metrics for direct target-d...  [Target-Disease]  \n",
       "4  Overall metrics for direct target-disease asso...  [Target-Disease]  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "\n",
    "# specifying the spreadsheet in the file 26.03:\n",
    "gid='194385091'\n",
    "\n",
    "# Curation of all columns from all OpenTargets output datasets:\n",
    "curation = f'https://docs.google.com/spreadsheets/d/17uyBBAu62ugRApNftu8hILuv_Qhh6_VfOUj91Tu-_R0/export?format=tsv&gid={gid}'\n",
    "# curation = 'annotated_new_evidence_columns.csv'\n",
    "# Folder to save the resulting curation file:\n",
    "asset_folder = '../src/ot_croissant/assets/'\n",
    "\n",
    "# Reading table:\n",
    "curation_table = (\n",
    "    pd.read_csv(curation, sep='\\t')\n",
    "    .assign(\n",
    "        tags=lambda df: df.tags.fillna('').apply(lambda x: x.split('|') if x else []),\n",
    "        key=lambda df: df.key.str.split('|'),\n",
    "    )\n",
    "    .sort_values(['id'])\n",
    ")\n",
    "curation_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = [\n",
    "  \"id\", \"nice_name\", \"key\", \"tags\", \"description\"\n",
    "]\n",
    "\n",
    "\n",
    "with open(f'{asset_folder}/distribution.json', 'w') as f:\n",
    "    json.dump(curation_table[columns].to_dict(orient='records'), f, indent=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ot-croissant",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
