{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating record set curation file\n",
    "\n",
    "- Fields in recordsets are annotated as a manual effort done by multiple people.\n",
    "- Curation was collected in a Google spreadsheet table.\n",
    "- The table has columns to make it easier for the curators to annotated also, also provides extra annotation for croissant ingestion\n",
    "\n",
    "## Columns\n",
    "\n",
    "- `dataset_name` - name of the dataset for curators.\n",
    "- `field_id` - identifier of the field for croissant.\n",
    "- `column_name` - just the label of the field for curators.\n",
    "- `column_description` - Curators annotate fields with description\n",
    "- `foreign_key` - curators add `field_id` of the foreign field\n",
    "- `bioregistry_prefix` - if data in a column comes from a database in bioregisty, annotate\n",
    "- `example` - helps curatos \n",
    "\n",
    "## Process\n",
    "\n",
    "1. Fetch curation table from Google\n",
    "2. Composing description\n",
    "3. Iterating over column and building output\n",
    "4. Save curation as json."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                          dataset_name  \\\n",
      "1184  association_by_datasource_direct   \n",
      "1185  association_by_datasource_direct   \n",
      "1186  association_by_datasource_direct   \n",
      "1193  association_by_datasource_direct   \n",
      "1182  association_by_datasource_direct   \n",
      "\n",
      "                                               field_id       column_name  \\\n",
      "1184   association_by_datasource_direct/aggregationType   aggregationType   \n",
      "1185  association_by_datasource_direct/aggregationValue  aggregationValue   \n",
      "1186  association_by_datasource_direct/associationScore  associationScore   \n",
      "1193    association_by_datasource_direct/currentNovelty    currentNovelty   \n",
      "1182         association_by_datasource_direct/diseaseId         diseaseId   \n",
      "\n",
      "                                     column_description foreign_key  \\\n",
      "1184  Evidence column used to aggregate data: dataso...        <NA>   \n",
      "1185  Datasource identifier of the group: eg. gwas_c...        <NA>   \n",
      "1186  Current disease/target association score of ev...        <NA>   \n",
      "1193          Novelty of the disease/target association        <NA>   \n",
      "1182                    Open Targets disease identifier  disease/id   \n",
      "\n",
      "     bioregistry_prefix  \n",
      "1184                NaN  \n",
      "1185                NaN  \n",
      "1186                NaN  \n",
      "1193                NaN  \n",
      "1182                NaN  \n",
      "1254\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "# specifying the spreadsheet in the file 26.03:\n",
    "gid='328767591'\n",
    "\n",
    "# Curation of all columns from all OpenTargets output datasets:\n",
    "curation = f'https://docs.google.com/spreadsheets/d/132SKHMoaJePu4nTlBnQwfaz3dhfJiKmJUujfYkzXMdI/export?format=tsv&gid={gid}'\n",
    "# curation = 'annotated_new_evidence_columns.csv'\n",
    "# Folder to save the resulting curation file:\n",
    "asset_folder = '../src/ot_croissant/assets/'\n",
    "\n",
    "# Reading table:\n",
    "curation_table = (\n",
    "    pd.read_csv(curation, sep='\\t')\n",
    "    .astype(\n",
    "        {\n",
    "            'column_description': pd.StringDtype(),\n",
    "            'foreign_key': pd.StringDtype(),\n",
    "        }\n",
    "    )\n",
    "    .assign(\n",
    "        column_description = lambda df: df.column_description.str.strip()\n",
    "    )\n",
    "    .sort_values(['dataset_name', 'field_id'])\n",
    ")\n",
    "print(curation_table.head())\n",
    "print(len(curation_table))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collection of curated dataset:\n",
    "curation_json = []\n",
    "\n",
    "# Composing description:\n",
    "def compose_description(row: pd.Series) -> str:\n",
    "    \"\"\"\n",
    "    Composes the description of a column based on the bioregistry prefix and the column description.\n",
    "    If the bioregistry prefix is not available, it returns the column description as is.\n",
    "\n",
    "    Args:\n",
    "        row (pd.Series): A row from the curation table.\n",
    "\n",
    "    Returns:\n",
    "        str: The composed description.\n",
    "    \"\"\"\n",
    "    # If the bioregistry prefix is not available, return the column description as is:\n",
    "    description = (\n",
    "        row['column_description']\n",
    "        if pd.isna(row['bioregistry_prefix'])\n",
    "        else f\"{row['column_description']} [bioregistry:{row['bioregistry_prefix'].lower()}]\"\n",
    "    )\n",
    "\n",
    "    return description\n",
    "\n",
    "# Iterating over the rows of the curation table:\n",
    "for _, row in curation_table.iterrows():\n",
    "    # If the column description is not available, skip the row:\n",
    "    if pd.isna(row['column_description']):\n",
    "        continue\n",
    "\n",
    "    # Adding curation to the dictionary:\n",
    "    data = {\n",
    "        'id': row['field_id'],\n",
    "        'description': compose_description(row)\n",
    "    }\n",
    "\n",
    "    # If the foreign key is available, add it to the dictionary:\n",
    "    if not pd.isna(row['foreign_key']):\n",
    "        data['foreign_key'] = row['foreign_key']\n",
    "\n",
    "    # If the bioregistry prefix is available, add it to the dictionary:\n",
    "    curation_json.append(data)\n",
    "\n",
    "# Saving the curation to a JSON file:\n",
    "with open(f'{asset_folder}/recordset.json', 'w') as f:\n",
    "    json.dump(curation_json, f, indent=2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>field_id</th>\n",
       "      <th>column_name</th>\n",
       "      <th>column_description</th>\n",
       "      <th>foreign_key</th>\n",
       "      <th>bioregistry_prefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>association_by_datasource_direct</td>\n",
       "      <td>association_by_datasource_direct/targetId</td>\n",
       "      <td>targetId</td>\n",
       "      <td>Unique identifier for the target</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1195</th>\n",
       "      <td>association_by_datasource_indirect</td>\n",
       "      <td>association_by_datasource_indirect/targetId</td>\n",
       "      <td>targetId</td>\n",
       "      <td>Unique identifier for the target</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207</th>\n",
       "      <td>association_by_datatype_direct</td>\n",
       "      <td>association_by_datatype_direct/targetId</td>\n",
       "      <td>targetId</td>\n",
       "      <td>Unique identifier for the target</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1219</th>\n",
       "      <td>association_by_datatype_indirect</td>\n",
       "      <td>association_by_datatype_indirect/targetId</td>\n",
       "      <td>targetId</td>\n",
       "      <td>Unique identifier for the target</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1231</th>\n",
       "      <td>association_overall_direct</td>\n",
       "      <td>association_overall_direct/targetId</td>\n",
       "      <td>targetId</td>\n",
       "      <td>Unique identifier for the target</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1151</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/transcriptConsequences/targetId</td>\n",
       "      <td>targetId</td>\n",
       "      <td>Open Target target identifier of the transcript</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1152</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/transcriptConsequences/transcriptId</td>\n",
       "      <td>transcriptId</td>\n",
       "      <td>Ensembl transcript identifier</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1154</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/transcriptConsequences/uniprotAccessions</td>\n",
       "      <td>uniprotAccessions</td>\n",
       "      <td>Uniprot identifiers of the gene product</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>UNIPROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1155</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/transcriptConsequences/variantFunction...</td>\n",
       "      <td>variantFunctionalConsequenceIds</td>\n",
       "      <td>The sequence ontology identifier of the conseq...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1163</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/variantEffect/targetId</td>\n",
       "      <td>targetId</td>\n",
       "      <td>Target identifier on which the variant effect ...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            dataset_name  \\\n",
       "1183    association_by_datasource_direct   \n",
       "1195  association_by_datasource_indirect   \n",
       "1207      association_by_datatype_direct   \n",
       "1219    association_by_datatype_indirect   \n",
       "1231          association_overall_direct   \n",
       "...                                  ...   \n",
       "1151                             variant   \n",
       "1152                             variant   \n",
       "1154                             variant   \n",
       "1155                             variant   \n",
       "1163                             variant   \n",
       "\n",
       "                                               field_id  \\\n",
       "1183          association_by_datasource_direct/targetId   \n",
       "1195        association_by_datasource_indirect/targetId   \n",
       "1207            association_by_datatype_direct/targetId   \n",
       "1219          association_by_datatype_indirect/targetId   \n",
       "1231                association_overall_direct/targetId   \n",
       "...                                                 ...   \n",
       "1151            variant/transcriptConsequences/targetId   \n",
       "1152        variant/transcriptConsequences/transcriptId   \n",
       "1154   variant/transcriptConsequences/uniprotAccessions   \n",
       "1155  variant/transcriptConsequences/variantFunction...   \n",
       "1163                     variant/variantEffect/targetId   \n",
       "\n",
       "                          column_name  \\\n",
       "1183                         targetId   \n",
       "1195                         targetId   \n",
       "1207                         targetId   \n",
       "1219                         targetId   \n",
       "1231                         targetId   \n",
       "...                               ...   \n",
       "1151                         targetId   \n",
       "1152                     transcriptId   \n",
       "1154                uniprotAccessions   \n",
       "1155  variantFunctionalConsequenceIds   \n",
       "1163                         targetId   \n",
       "\n",
       "                                     column_description foreign_key  \\\n",
       "1183                   Unique identifier for the target   target/id   \n",
       "1195                   Unique identifier for the target   target/id   \n",
       "1207                   Unique identifier for the target   target/id   \n",
       "1219                   Unique identifier for the target   target/id   \n",
       "1231                   Unique identifier for the target   target/id   \n",
       "...                                                 ...         ...   \n",
       "1151    Open Target target identifier of the transcript   target/id   \n",
       "1152                      Ensembl transcript identifier        <NA>   \n",
       "1154            Uniprot identifiers of the gene product        <NA>   \n",
       "1155  The sequence ontology identifier of the conseq...       so/id   \n",
       "1163  Target identifier on which the variant effect ...   target/id   \n",
       "\n",
       "     bioregistry_prefix  \n",
       "1183            ENSEMBL  \n",
       "1195            ENSEMBL  \n",
       "1207            ENSEMBL  \n",
       "1219            ENSEMBL  \n",
       "1231            ENSEMBL  \n",
       "...                 ...  \n",
       "1151            ENSEMBL  \n",
       "1152            ENSEMBL  \n",
       "1154            UNIPROT  \n",
       "1155                 SO  \n",
       "1163            ENSEMBL  \n",
       "\n",
       "[71 rows x 6 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Columns with bioregistry prefix:\n",
    "curation_table.loc[curation_table.bioregistry_prefix.notna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>field_id</th>\n",
       "      <th>column_name</th>\n",
       "      <th>column_description</th>\n",
       "      <th>foreign_key</th>\n",
       "      <th>bioregistry_prefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/tissues/protein/cell_type/level</td>\n",
       "      <td>level</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/tissues/protein/cell_type/reliability</td>\n",
       "      <td>reliability</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/tissues/protein/reliability</td>\n",
       "      <td>reliability</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    dataset_name                                          field_id  \\\n",
       "716   expression        expression/tissues/protein/cell_type/level   \n",
       "718   expression  expression/tissues/protein/cell_type/reliability   \n",
       "720   expression            expression/tissues/protein/reliability   \n",
       "\n",
       "     column_name column_description foreign_key bioregistry_prefix  \n",
       "716        level               <NA>        <NA>                NaN  \n",
       "718  reliability               <NA>        <NA>                NaN  \n",
       "720  reliability               <NA>        <NA>                NaN  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fields without description:\n",
    "curation_table.loc[curation_table.column_description.isna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate distributions JSON\n",
    "\n",
    "\n",
    "Creat distribution json based on curation table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>nice_name</th>\n",
       "      <th>key</th>\n",
       "      <th>description</th>\n",
       "      <th>tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>association_by_datasource_direct</td>\n",
       "      <td>Associations - direct (by data source)</td>\n",
       "      <td>[association_by_datasource_direct/diseaseId, a...</td>\n",
       "      <td>Data source-specific metrics for direct target...</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>association_by_datasource_indirect</td>\n",
       "      <td>Associations - indirect (by data source)</td>\n",
       "      <td>[association_by_datasource_indirect/diseaseId,...</td>\n",
       "      <td>Data source-specific metrics for indirect targ...</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>association_by_datatype_direct</td>\n",
       "      <td>Associations - direct (by data type)</td>\n",
       "      <td>[association_by_datatype_direct/diseaseId, ass...</td>\n",
       "      <td>Data type-specific metrics for direct target-d...</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>association_by_datatype_indirect</td>\n",
       "      <td>Associations - indirect (by data type)</td>\n",
       "      <td>[association_by_datatype_indirect/diseaseId, a...</td>\n",
       "      <td>Data type-specific metrics for indirect target...</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>association_overall_direct</td>\n",
       "      <td>Associations - direct (overall score)</td>\n",
       "      <td>[association_overall_direct/diseaseId, associa...</td>\n",
       "      <td>Overall metrics for direct target-disease asso...</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   id  \\\n",
       "0    association_by_datasource_direct   \n",
       "1  association_by_datasource_indirect   \n",
       "2      association_by_datatype_direct   \n",
       "3    association_by_datatype_indirect   \n",
       "4          association_overall_direct   \n",
       "\n",
       "                                  nice_name  \\\n",
       "0    Associations - direct (by data source)   \n",
       "1  Associations - indirect (by data source)   \n",
       "2      Associations - direct (by data type)   \n",
       "3    Associations - indirect (by data type)   \n",
       "4     Associations - direct (overall score)   \n",
       "\n",
       "                                                 key  \\\n",
       "0  [association_by_datasource_direct/diseaseId, a...   \n",
       "1  [association_by_datasource_indirect/diseaseId,...   \n",
       "2  [association_by_datatype_direct/diseaseId, ass...   \n",
       "3  [association_by_datatype_indirect/diseaseId, a...   \n",
       "4  [association_overall_direct/diseaseId, associa...   \n",
       "\n",
       "                                         description              tags  \n",
       "0  Data source-specific metrics for direct target...  [Target-Disease]  \n",
       "1  Data source-specific metrics for indirect targ...  [Target-Disease]  \n",
       "2  Data type-specific metrics for direct target-d...  [Target-Disease]  \n",
       "3  Data type-specific metrics for indirect target...  [Target-Disease]  \n",
       "4  Overall metrics for direct target-disease asso...  [Target-Disease]  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "\n",
    "# specifying the spreadsheet in the file 26.03:\n",
    "gid='194385091'\n",
    "\n",
    "# Curation of all columns from all OpenTargets output datasets:\n",
    "curation = f'https://docs.google.com/spreadsheets/d/17uyBBAu62ugRApNftu8hILuv_Qhh6_VfOUj91Tu-_R0/export?format=tsv&gid={gid}'\n",
    "# curation = 'annotated_new_evidence_columns.csv'\n",
    "# Folder to save the resulting curation file:\n",
    "asset_folder = '../src/ot_croissant/assets/'\n",
    "\n",
    "# Reading table:\n",
    "curation_table = (\n",
    "    pd.read_csv(curation, sep='\\t')\n",
    "    .assign(\n",
    "        tags=lambda df: df.tags.fillna('').apply(lambda x: x.split('|') if x else []),\n",
    "        key=lambda df: df.key.str.split('|'),\n",
    "    )\n",
    "    .sort_values(['id'])\n",
    ")\n",
    "curation_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = [\n",
    "  \"id\", \"nice_name\", \"key\", \"tags\", \"description\"\n",
    "]\n",
    "\n",
    "\n",
    "with open(f'{asset_folder}/distribution.json', 'w') as f:\n",
    "    json.dump(curation_table[columns].to_dict(orient='records'), f, indent=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding representation of new datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "26/02/04 11:43:11 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "72\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>field_id</th>\n",
       "      <th>column_name</th>\n",
       "      <th>field_join</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>association_by_datasource_direct</td>\n",
       "      <td>association_by_datasource_direct/diseaseId</td>\n",
       "      <td>diseaseId</td>\n",
       "      <td>diseaseId</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>association_by_datasource_direct</td>\n",
       "      <td>association_by_datasource_direct/targetId</td>\n",
       "      <td>targetId</td>\n",
       "      <td>targetId</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>association_by_datasource_direct</td>\n",
       "      <td>association_by_datasource_direct/aggregationType</td>\n",
       "      <td>aggregationType</td>\n",
       "      <td>aggregationType</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>association_by_datasource_direct</td>\n",
       "      <td>association_by_datasource_direct/aggregationValue</td>\n",
       "      <td>aggregationValue</td>\n",
       "      <td>aggregationValue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>association_by_datasource_direct</td>\n",
       "      <td>association_by_datasource_direct/associationScore</td>\n",
       "      <td>associationScore</td>\n",
       "      <td>associationScore</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       dataset_name  \\\n",
       "0  association_by_datasource_direct   \n",
       "1  association_by_datasource_direct   \n",
       "2  association_by_datasource_direct   \n",
       "3  association_by_datasource_direct   \n",
       "4  association_by_datasource_direct   \n",
       "\n",
       "                                            field_id       column_name  \\\n",
       "0         association_by_datasource_direct/diseaseId         diseaseId   \n",
       "1          association_by_datasource_direct/targetId          targetId   \n",
       "2   association_by_datasource_direct/aggregationType   aggregationType   \n",
       "3  association_by_datasource_direct/aggregationValue  aggregationValue   \n",
       "4  association_by_datasource_direct/associationScore  associationScore   \n",
       "\n",
       "         field_join  \n",
       "0         diseaseId  \n",
       "1          targetId  \n",
       "2   aggregationType  \n",
       "3  aggregationValue  \n",
       "4  associationScore  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pyspark.sql import SparkSession, functions as f\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "\n",
    "# Processing a field in the spark schema:\n",
    "def process_fields(\n",
    "    fields: dict[str, str], \n",
    "    parent: str, \n",
    "    dataset_name: str,\n",
    "    schema: list[dict[str,str]] | None = None,\n",
    ") -> list[dict[str, str]]:\n",
    "    if schema is None:\n",
    "        schema = []\n",
    "\n",
    "    # Iterating over all fields of the schema:\n",
    "    for field in fields.get('fields'):\n",
    "\n",
    "        name = field.get('name')\n",
    "        # Recurse for nested structs:\n",
    "        if isinstance(field['type'], dict):\n",
    "            if field['type']['type'] == 'struct':\n",
    "                # Call self:\n",
    "                schema = schema + process_fields(\n",
    "                    field['type'], \n",
    "                    f'{parent}/{name}',\n",
    "                    dataset_name,\n",
    "                    [],\n",
    "                )\n",
    "            elif isinstance(field['type']['elementType'], dict):\n",
    "                # Call self:\n",
    "                schema = schema + process_fields(\n",
    "                    field[\"type\"]['elementType'], \n",
    "                    f'{parent}/{name}',\n",
    "                    dataset_name,\n",
    "                    [],\n",
    "                )\n",
    "        # Capture atomic column types:\n",
    "        schema.append(\n",
    "            {\n",
    "                'dataset_name': dataset_name,\n",
    "                'field_id': f'{parent}/{name}',\n",
    "                'column_name': name,\n",
    "            }\n",
    "        )\n",
    "    # Return captured schema:\n",
    "    return schema\n",
    "\n",
    "\n",
    "def generate_schema_representation(\n",
    "    dataset_location: str, output: str,\n",
    ") -> pd.DataFrame:\n",
    "\n",
    "    # Read dataset:\n",
    "    try:\n",
    "        df = spark.read.parquet(f'{dataset_location}/{output}')\n",
    "    except:\n",
    "        print(f'missing dataset: {dataset_location}/{output}')\n",
    "        return None\n",
    "\n",
    "    # Get its schema:\n",
    "    data = json.loads(df.schema.json())\n",
    "\n",
    "    # We iterate over the schema and collect field names that we will be able to annotate on the spreadsheet:\n",
    "    schema = process_fields(\n",
    "        data, \n",
    "        output,\n",
    "        output\n",
    "    )\n",
    "\n",
    "    # Process and return schema representation:\n",
    "    return (\n",
    "        pd.DataFrame(schema)\n",
    "        .assign(\n",
    "            field_join = lambda x: x.field_id.str.extract(r'\\/(.+)$')\n",
    "        )\n",
    "    )\n",
    "\n",
    "dataset_location = '/Users/dsuveges/project_data/releases/25.12/output'\n",
    "datasets = [\n",
    "    'association_by_datasource_direct', \n",
    "    'association_by_datasource_indirect', \n",
    "    'association_by_datatype_direct', \n",
    "    'association_by_datatype_indirect', \n",
    "    'association_overall_direct', \n",
    "    'association_overall_indirect'\n",
    "]\n",
    "\n",
    "\n",
    "# generate_schema_representation(**dataset)    \n",
    "new_evidence_collated = pd.concat([\n",
    "    generate_schema_representation(dataset_location, dataset)\n",
    "    for dataset in \n",
    "    datasets\n",
    "])\n",
    "\n",
    "print(len(new_evidence_collated))\n",
    "new_evidence_collated.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_evidence_collated.to_csv(\n",
    "    'associations.tsv', sep=',', index = False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ot-croissant",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
