{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating record set curation file\n",
    "\n",
    "- Fields in recordsets are annotated as a manual effort done by multiple people.\n",
    "- Curation was collected in a Google spreadsheet table.\n",
    "- The table has columns to make it easier for the curators to annotated also, also provides extra annotation for croissant ingestion\n",
    "\n",
    "## Columns\n",
    "\n",
    "- `dataset_name` - name of the dataset for curators.\n",
    "- `field_id` - identifier of the field for croissant.\n",
    "- `column_name` - just the label of the field for curators.\n",
    "- `column_description` - Curators annotate fields with description\n",
    "- `foreign_key` - curators add `field_id` of the foreign field\n",
    "- `bioregistry_prefix` - if data in a column comes from a database in bioregisty, annotate\n",
    "- `example` - helps curatos \n",
    "\n",
    "## Process\n",
    "\n",
    "1. Fetch curation table from Google\n",
    "2. Composing description\n",
    "3. Iterating over column and building output\n",
    "4. Save curation as json."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>field_id</th>\n",
       "      <th>column_name</th>\n",
       "      <th>column_description</th>\n",
       "      <th>foreign_key</th>\n",
       "      <th>bioregistry_prefix</th>\n",
       "      <th>Example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>disease_phenotype</td>\n",
       "      <td>disease_phenotype/disease</td>\n",
       "      <td>disease</td>\n",
       "      <td>Disease identifier</td>\n",
       "      <td>disease/id</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO_0800026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>disease_phenotype</td>\n",
       "      <td>disease_phenotype/phenotype</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>The phenotype linked to the disease.</td>\n",
       "      <td>disease/id</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>disease_phenotype</td>\n",
       "      <td>disease_phenotype/evidence</td>\n",
       "      <td>evidence</td>\n",
       "      <td>A container for all evidence-related attribute...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>disease_phenotype</td>\n",
       "      <td>disease_phenotype/evidence/aspect</td>\n",
       "      <td>aspect</td>\n",
       "      <td>The category of biological information being p...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>disease_phenotype</td>\n",
       "      <td>disease_phenotype/evidence/bioCuration</td>\n",
       "      <td>bioCuration</td>\n",
       "      <td>Indicates whether the evidence has been manual...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HPO:probinson[2021-09-23];HPO:probinson[2021-0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dataset_name                                field_id  column_name  \\\n",
       "0  disease_phenotype               disease_phenotype/disease      disease   \n",
       "1  disease_phenotype             disease_phenotype/phenotype    phenotype   \n",
       "2  disease_phenotype              disease_phenotype/evidence     evidence   \n",
       "3  disease_phenotype       disease_phenotype/evidence/aspect       aspect   \n",
       "4  disease_phenotype  disease_phenotype/evidence/bioCuration  bioCuration   \n",
       "\n",
       "                                  column_description foreign_key  \\\n",
       "0                                 Disease identifier  disease/id   \n",
       "1               The phenotype linked to the disease.  disease/id   \n",
       "2  A container for all evidence-related attribute...        <NA>   \n",
       "3  The category of biological information being p...        <NA>   \n",
       "4  Indicates whether the evidence has been manual...        <NA>   \n",
       "\n",
       "  bioregistry_prefix                                            Example  \n",
       "0                NaN                                      MONDO_0800026  \n",
       "1                NaN                                                NaN  \n",
       "2                NaN                                                NaN  \n",
       "3                NaN                                                  C  \n",
       "4                NaN  HPO:probinson[2021-09-23];HPO:probinson[2021-0...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "# Curation of all columns from all OpenTargets output datasets:\n",
    "curation = 'https://docs.google.com/spreadsheets/d/132SKHMoaJePu4nTlBnQwfaz3dhfJiKmJUujfYkzXMdI/export?format=tsv&gid=179018892'\n",
    "\n",
    "# Folder to save the resulting curation file:\n",
    "asset_folder = '../src/ot_croissant/assets/'\n",
    "\n",
    "# Reading table:\n",
    "curation_table = (\n",
    "    pd.read_csv(curation, sep='\\t')\n",
    "    .astype(\n",
    "        {\n",
    "            'column_description': pd.StringDtype(),\n",
    "            'foreign_key': pd.StringDtype(),\n",
    "        }\n",
    "    )\n",
    ")\n",
    "curation_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collection of curated dataset:\n",
    "curation_json = []\n",
    "\n",
    "# Composing description:\n",
    "def compose_description(row: pd.Series) -> str:\n",
    "    \"\"\"\n",
    "    Composes the description of a column based on the bioregistry prefix and the column description.\n",
    "    If the bioregistry prefix is not available, it returns the column description as is.\n",
    "\n",
    "    Args:\n",
    "        row (pd.Series): A row from the curation table.\n",
    "    \n",
    "    Returns:\n",
    "        str: The composed description.\n",
    "    \"\"\"\n",
    "    # If the bioregistry prefix is not available, return the column description as is:\n",
    "    description = (\n",
    "        row['column_description']\n",
    "        if pd.isna(row['bioregistry_prefix'])\n",
    "        else f\"{row['column_description']} [bioregistry:{row['bioregistry_prefix'].lower()}]\"\n",
    "    )\n",
    "\n",
    "    return description\n",
    "\n",
    "# Iterating over the rows of the curation table:\n",
    "for _, row in curation_table.iterrows():\n",
    "    # If the column description is not available, skip the row:\n",
    "    if pd.isna(row['column_description']):\n",
    "        continue\n",
    "\n",
    "    # Adding curation to the dictionary:\n",
    "    data = {\n",
    "        'id': row['field_id'],\n",
    "        'description': compose_description(row)\n",
    "    }\n",
    "\n",
    "    # If the foreign key is available, add it to the dictionary:\n",
    "    if not pd.isna(row['foreign_key']):\n",
    "        data['foreign_key'] = row['foreign_key']\n",
    "\n",
    "    # If the bioregistry prefix is available, add it to the dictionary:\n",
    "    curation_json.append(data)\n",
    "\n",
    "# Saving the curation to a JSON file:\n",
    "with open(f'{asset_folder}/recordset.json', 'w') as f:\n",
    "    json.dump(curation_json, f, indent=2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>field_id</th>\n",
       "      <th>column_name</th>\n",
       "      <th>column_description</th>\n",
       "      <th>foreign_key</th>\n",
       "      <th>bioregistry_prefix</th>\n",
       "      <th>Example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>disease_phenotype</td>\n",
       "      <td>disease_phenotype/evidence/references</td>\n",
       "      <td>references</td>\n",
       "      <td>References or citations supporting the evidence.</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed</td>\n",
       "      <td>[PMID:14566559]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>mouse_phenotype</td>\n",
       "      <td>mouse_phenotype/biologicalModels/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Unique identifier for the biological model.</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:6140117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>mouse_phenotype</td>\n",
       "      <td>mouse_phenotype/biologicalModels/literature</td>\n",
       "      <td>literature</td>\n",
       "      <td>References related to the mouse model.</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed</td>\n",
       "      <td>[30949703]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>mouse_phenotype</td>\n",
       "      <td>mouse_phenotype/modelPhenotypeClasses/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Unique identifier for the phenotype class.</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MP</td>\n",
       "      <td>MP:0005389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>mouse_phenotype</td>\n",
       "      <td>mouse_phenotype/modelPhenotypeId</td>\n",
       "      <td>modelPhenotypeId</td>\n",
       "      <td>Identifier for the specific phenotype observed...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MP</td>\n",
       "      <td>MP:0005343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>mouse_phenotype</td>\n",
       "      <td>mouse_phenotype/targetInModelEnsemblId</td>\n",
       "      <td>targetInModelEnsemblId</td>\n",
       "      <td>Ensembl identifier for the target gene in the ...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>ENSMUSG00000087651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>mouse_phenotype</td>\n",
       "      <td>mouse_phenotype/targetInModelMgiId</td>\n",
       "      <td>targetInModelMgiId</td>\n",
       "      <td>MGI (Mouse Genome Informatics) identifier for ...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:1917034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>reactome</td>\n",
       "      <td>reactome/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Unique identifier for the Reactome pathway</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>reactome</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Ensembl human gene identifier for the expresse...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>ENSG00000071243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/tissues/efo_code</td>\n",
       "      <td>efo_code</td>\n",
       "      <td>Ontology ID of the biosample the expression da...</td>\n",
       "      <td>biosample/biosampleId</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>target</td>\n",
       "      <td>target/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Unique identifier for the target</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>target</td>\n",
       "      <td>target/go/id</td>\n",
       "      <td>id</td>\n",
       "      <td>GO identifier for the annotation</td>\n",
       "      <td>go/id</td>\n",
       "      <td>GO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>target</td>\n",
       "      <td>target/go/geneProduct</td>\n",
       "      <td>geneProduct</td>\n",
       "      <td>Gene product associated with the GO annotation</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>target</td>\n",
       "      <td>target/go/ecoId</td>\n",
       "      <td>ecoId</td>\n",
       "      <td>Evidence &amp; Conclusion Ontology (ECO) identifie...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ECO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>target</td>\n",
       "      <td>target/hallmarks/attributes/pmid</td>\n",
       "      <td>pmid</td>\n",
       "      <td>PubMed ID of the supporting literature for the...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>target</td>\n",
       "      <td>target/hallmarks/cancerHallmarks/pmid</td>\n",
       "      <td>pmid</td>\n",
       "      <td>PubMed ID of the supporting literature for the...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>target</td>\n",
       "      <td>target/subcellularLocations/termSL</td>\n",
       "      <td>termSL</td>\n",
       "      <td>Subcellular location term identifier from Swis...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>SL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>go</td>\n",
       "      <td>go/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Gene ontology term identifier</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>go</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>so</td>\n",
       "      <td>so/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Sequence ontology term identifier</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>so</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>biosample</td>\n",
       "      <td>biosample/biosampleId</td>\n",
       "      <td>biosampleId</td>\n",
       "      <td>Unique identifier for the biosample</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391</th>\n",
       "      <td>evidence</td>\n",
       "      <td>evidence/ancestryId</td>\n",
       "      <td>ancestryId</td>\n",
       "      <td>Identifier of the ancestry in the HANCESTRO on...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>HANCESTRO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>evidence</td>\n",
       "      <td>evidence/biomarkers/geneExpression/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Gene Ontology (GO) identifiers of regulation o...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>GO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>evidence</td>\n",
       "      <td>evidence/biomarkers/geneticVariation/functiona...</td>\n",
       "      <td>functionalConsequenceId</td>\n",
       "      <td>Functional consequene identifier of the varian...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>evidence</td>\n",
       "      <td>evidence/diseaseCellLines/tissueId</td>\n",
       "      <td>tissueId</td>\n",
       "      <td>Anatomical identifier of the sampled organ/tissue</td>\n",
       "      <td>biosample/biosampleId</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>evidence</td>\n",
       "      <td>evidence/diseaseModelAssociatedHumanPhenotypes/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Human phenotype ontology identifier</td>\n",
       "      <td>disease_hpo/id</td>\n",
       "      <td>HP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>evidence</td>\n",
       "      <td>evidence/diseaseModelAssociatedModelPhenotypes/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Mouse phenotype ontology identifier</td>\n",
       "      <td>mouse_phenotype/modelPhenotypeId</td>\n",
       "      <td>MP</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>evidence</td>\n",
       "      <td>evidence/drugId</td>\n",
       "      <td>drugId</td>\n",
       "      <td>Open Targets drug identifier</td>\n",
       "      <td>drug_molecule/id</td>\n",
       "      <td>CHEMBL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>evidence</td>\n",
       "      <td>evidence/mutatedSamples/functionalConsequenceId</td>\n",
       "      <td>functionalConsequenceId</td>\n",
       "      <td>Functional consequene identifier of the pooled...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>evidence</td>\n",
       "      <td>evidence/pathways/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Reactome pathway identifier</td>\n",
       "      <td>reactome/id</td>\n",
       "      <td>reactome</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>evidence</td>\n",
       "      <td>evidence/targetInModelEnsemblId</td>\n",
       "      <td>targetInModelEnsemblId</td>\n",
       "      <td>Target Ensembl ID in animal model</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>evidence</td>\n",
       "      <td>evidence/targetInModelMgiId</td>\n",
       "      <td>targetInModelMgiId</td>\n",
       "      <td>Target MGI ID in animal model</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MGI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>505</th>\n",
       "      <td>evidence</td>\n",
       "      <td>evidence/variantFunctionalConsequenceId</td>\n",
       "      <td>variantFunctionalConsequenceId</td>\n",
       "      <td>Sequence ontology (SO) identifier of the funct...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>pharmacogenomics</td>\n",
       "      <td>pharmacogenomics/variantFunctionalConsequenceId</td>\n",
       "      <td>variantFunctionalConsequenceId</td>\n",
       "      <td>The sequence ontology identifier of the conseq...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548</th>\n",
       "      <td>target_essentiality</td>\n",
       "      <td>target_essentiality/id</td>\n",
       "      <td>id</td>\n",
       "      <td>Target identifier the essentiality was measure...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>559</th>\n",
       "      <td>target_essentiality</td>\n",
       "      <td>target_essentiality/geneEssentiality/depMapEss...</td>\n",
       "      <td>tissueId</td>\n",
       "      <td>Identifier of the tissue from where the cells ...</td>\n",
       "      <td>biosample/biosampleId</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/targetB</td>\n",
       "      <td>targetB</td>\n",
       "      <td>Open Targets target identifier of the second m...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/participantDetectionMetho...</td>\n",
       "      <td>miIdentifier</td>\n",
       "      <td>Molecular Interactions (MI) identifier for par...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>608</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/interactionTypeMiIdentifier</td>\n",
       "      <td>interactionTypeMiIdentifier</td>\n",
       "      <td>Molecular Interactions (MI) identifier for the...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>609</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/expansionMethodMiIdentifier</td>\n",
       "      <td>expansionMethodMiIdentifier</td>\n",
       "      <td>Molecular Interactions (MI) identifier for the...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/participantDetectionMetho...</td>\n",
       "      <td>miIdentifier</td>\n",
       "      <td>Molecular Interactions (MI) identifier for par...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/targetA</td>\n",
       "      <td>targetA</td>\n",
       "      <td>Open Targets target identifier of the second m...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>633</th>\n",
       "      <td>interaction_evidence</td>\n",
       "      <td>interaction_evidence/interactionDetectionMetho...</td>\n",
       "      <td>interactionDetectionMethodMiIdentifier</td>\n",
       "      <td>Molecular Interactions (MI) identifier for the...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/variantEffect/targetId</td>\n",
       "      <td>targetId</td>\n",
       "      <td>The target identifier on which the variant eff...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/mostSevereConsequenceId</td>\n",
       "      <td>mostSevereConsequenceId</td>\n",
       "      <td>The sequence ontology identifier of the most s...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/transcriptConsequences/variantFunction...</td>\n",
       "      <td>variantFunctionalConsequenceIds</td>\n",
       "      <td>The sequence ontology identifier of the conseq...</td>\n",
       "      <td>so/id</td>\n",
       "      <td>SO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>656</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/transcriptConsequences/uniprotAccessions</td>\n",
       "      <td>uniprotAccessions</td>\n",
       "      <td>List of Uniprot identifiers of the gene product</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>UNIPROT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/transcriptConsequences/targetId</td>\n",
       "      <td>targetId</td>\n",
       "      <td>Open Target target identifier of the transcript</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>672</th>\n",
       "      <td>variant</td>\n",
       "      <td>variant/transcriptConsequences/transcriptId</td>\n",
       "      <td>transcriptId</td>\n",
       "      <td>The Ensembl transcript identifier</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683</th>\n",
       "      <td>interaction</td>\n",
       "      <td>interaction/targetA</td>\n",
       "      <td>targetA</td>\n",
       "      <td>Open Targets target identifier of the first mo...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>686</th>\n",
       "      <td>interaction</td>\n",
       "      <td>interaction/targetB</td>\n",
       "      <td>targetB</td>\n",
       "      <td>Open Targets target identifier of the second m...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>study</td>\n",
       "      <td>study/geneId</td>\n",
       "      <td>geneId</td>\n",
       "      <td>Ensembl identifier of the gene representing a ...</td>\n",
       "      <td>target/id</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>study</td>\n",
       "      <td>study/pubmedId</td>\n",
       "      <td>pubmedId</td>\n",
       "      <td>PubMed identifier of the publication hat refer...</td>\n",
       "      <td>literature/pmid</td>\n",
       "      <td>PUBMED</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>study</td>\n",
       "      <td>study/biosampleId</td>\n",
       "      <td>biosampleId</td>\n",
       "      <td>Biosample identifier that maps to the biosampl...</td>\n",
       "      <td>biosample/biosampleId</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             dataset_name                                           field_id  \\\n",
       "14      disease_phenotype              disease_phenotype/evidence/references   \n",
       "20        mouse_phenotype                mouse_phenotype/biologicalModels/id   \n",
       "21        mouse_phenotype        mouse_phenotype/biologicalModels/literature   \n",
       "23        mouse_phenotype           mouse_phenotype/modelPhenotypeClasses/id   \n",
       "25        mouse_phenotype                   mouse_phenotype/modelPhenotypeId   \n",
       "29        mouse_phenotype             mouse_phenotype/targetInModelEnsemblId   \n",
       "30        mouse_phenotype                 mouse_phenotype/targetInModelMgiId   \n",
       "48               reactome                                        reactome/id   \n",
       "55             expression                                      expression/id   \n",
       "57             expression                        expression/tissues/efo_code   \n",
       "83                 target                                          target/id   \n",
       "102                target                                       target/go/id   \n",
       "106                target                              target/go/geneProduct   \n",
       "107                target                                    target/go/ecoId   \n",
       "110                target                   target/hallmarks/attributes/pmid   \n",
       "114                target              target/hallmarks/cancerHallmarks/pmid   \n",
       "131                target                 target/subcellularLocations/termSL   \n",
       "329                    go                                              go/id   \n",
       "331                    so                                              so/id   \n",
       "349             biosample                              biosample/biosampleId   \n",
       "391              evidence                                evidence/ancestryId   \n",
       "409              evidence              evidence/biomarkers/geneExpression/id   \n",
       "412              evidence  evidence/biomarkers/geneticVariation/functiona...   \n",
       "433              evidence                 evidence/diseaseCellLines/tissueId   \n",
       "438              evidence  evidence/diseaseModelAssociatedHumanPhenotypes/id   \n",
       "441              evidence  evidence/diseaseModelAssociatedModelPhenotypes/id   \n",
       "444              evidence                                    evidence/drugId   \n",
       "456              evidence    evidence/mutatedSamples/functionalConsequenceId   \n",
       "466              evidence                               evidence/pathways/id   \n",
       "496              evidence                    evidence/targetInModelEnsemblId   \n",
       "497              evidence                        evidence/targetInModelMgiId   \n",
       "505              evidence            evidence/variantFunctionalConsequenceId   \n",
       "541      pharmacogenomics    pharmacogenomics/variantFunctionalConsequenceId   \n",
       "548   target_essentiality                             target_essentiality/id   \n",
       "559   target_essentiality  target_essentiality/geneEssentiality/depMapEss...   \n",
       "596  interaction_evidence                       interaction_evidence/targetB   \n",
       "599  interaction_evidence  interaction_evidence/participantDetectionMetho...   \n",
       "608  interaction_evidence   interaction_evidence/interactionTypeMiIdentifier   \n",
       "609  interaction_evidence   interaction_evidence/expansionMethodMiIdentifier   \n",
       "628  interaction_evidence  interaction_evidence/participantDetectionMetho...   \n",
       "632  interaction_evidence                       interaction_evidence/targetA   \n",
       "633  interaction_evidence  interaction_evidence/interactionDetectionMetho...   \n",
       "650               variant                     variant/variantEffect/targetId   \n",
       "652               variant                    variant/mostSevereConsequenceId   \n",
       "654               variant  variant/transcriptConsequences/variantFunction...   \n",
       "656               variant   variant/transcriptConsequences/uniprotAccessions   \n",
       "663               variant            variant/transcriptConsequences/targetId   \n",
       "672               variant        variant/transcriptConsequences/transcriptId   \n",
       "683           interaction                                interaction/targetA   \n",
       "686           interaction                                interaction/targetB   \n",
       "700                 study                                       study/geneId   \n",
       "706                 study                                     study/pubmedId   \n",
       "736                 study                                  study/biosampleId   \n",
       "\n",
       "                                column_name  \\\n",
       "14                               references   \n",
       "20                                       id   \n",
       "21                               literature   \n",
       "23                                       id   \n",
       "25                         modelPhenotypeId   \n",
       "29                   targetInModelEnsemblId   \n",
       "30                       targetInModelMgiId   \n",
       "48                                       id   \n",
       "55                                       id   \n",
       "57                                 efo_code   \n",
       "83                                       id   \n",
       "102                                      id   \n",
       "106                             geneProduct   \n",
       "107                                   ecoId   \n",
       "110                                    pmid   \n",
       "114                                    pmid   \n",
       "131                                  termSL   \n",
       "329                                      id   \n",
       "331                                      id   \n",
       "349                             biosampleId   \n",
       "391                              ancestryId   \n",
       "409                                      id   \n",
       "412                 functionalConsequenceId   \n",
       "433                                tissueId   \n",
       "438                                      id   \n",
       "441                                      id   \n",
       "444                                  drugId   \n",
       "456                 functionalConsequenceId   \n",
       "466                                      id   \n",
       "496                  targetInModelEnsemblId   \n",
       "497                      targetInModelMgiId   \n",
       "505          variantFunctionalConsequenceId   \n",
       "541          variantFunctionalConsequenceId   \n",
       "548                                      id   \n",
       "559                                tissueId   \n",
       "596                                 targetB   \n",
       "599                            miIdentifier   \n",
       "608             interactionTypeMiIdentifier   \n",
       "609             expansionMethodMiIdentifier   \n",
       "628                            miIdentifier   \n",
       "632                                 targetA   \n",
       "633  interactionDetectionMethodMiIdentifier   \n",
       "650                                targetId   \n",
       "652                 mostSevereConsequenceId   \n",
       "654         variantFunctionalConsequenceIds   \n",
       "656                       uniprotAccessions   \n",
       "663                                targetId   \n",
       "672                            transcriptId   \n",
       "683                                 targetA   \n",
       "686                                 targetB   \n",
       "700                                  geneId   \n",
       "706                                pubmedId   \n",
       "736                             biosampleId   \n",
       "\n",
       "                                    column_description  \\\n",
       "14    References or citations supporting the evidence.   \n",
       "20         Unique identifier for the biological model.   \n",
       "21              References related to the mouse model.   \n",
       "23          Unique identifier for the phenotype class.   \n",
       "25   Identifier for the specific phenotype observed...   \n",
       "29   Ensembl identifier for the target gene in the ...   \n",
       "30   MGI (Mouse Genome Informatics) identifier for ...   \n",
       "48          Unique identifier for the Reactome pathway   \n",
       "55   Ensembl human gene identifier for the expresse...   \n",
       "57   Ontology ID of the biosample the expression da...   \n",
       "83                    Unique identifier for the target   \n",
       "102                   GO identifier for the annotation   \n",
       "106     Gene product associated with the GO annotation   \n",
       "107  Evidence & Conclusion Ontology (ECO) identifie...   \n",
       "110  PubMed ID of the supporting literature for the...   \n",
       "114  PubMed ID of the supporting literature for the...   \n",
       "131  Subcellular location term identifier from Swis...   \n",
       "329                      Gene ontology term identifier   \n",
       "331                  Sequence ontology term identifier   \n",
       "349                Unique identifier for the biosample   \n",
       "391  Identifier of the ancestry in the HANCESTRO on...   \n",
       "409  Gene Ontology (GO) identifiers of regulation o...   \n",
       "412  Functional consequene identifier of the varian...   \n",
       "433  Anatomical identifier of the sampled organ/tissue   \n",
       "438                Human phenotype ontology identifier   \n",
       "441                Mouse phenotype ontology identifier   \n",
       "444                       Open Targets drug identifier   \n",
       "456  Functional consequene identifier of the pooled...   \n",
       "466                        Reactome pathway identifier   \n",
       "496                  Target Ensembl ID in animal model   \n",
       "497                      Target MGI ID in animal model   \n",
       "505  Sequence ontology (SO) identifier of the funct...   \n",
       "541  The sequence ontology identifier of the conseq...   \n",
       "548  Target identifier the essentiality was measure...   \n",
       "559  Identifier of the tissue from where the cells ...   \n",
       "596  Open Targets target identifier of the second m...   \n",
       "599  Molecular Interactions (MI) identifier for par...   \n",
       "608  Molecular Interactions (MI) identifier for the...   \n",
       "609  Molecular Interactions (MI) identifier for the...   \n",
       "628  Molecular Interactions (MI) identifier for par...   \n",
       "632  Open Targets target identifier of the second m...   \n",
       "633  Molecular Interactions (MI) identifier for the...   \n",
       "650  The target identifier on which the variant eff...   \n",
       "652  The sequence ontology identifier of the most s...   \n",
       "654  The sequence ontology identifier of the conseq...   \n",
       "656    List of Uniprot identifiers of the gene product   \n",
       "663    Open Target target identifier of the transcript   \n",
       "672                  The Ensembl transcript identifier   \n",
       "683  Open Targets target identifier of the first mo...   \n",
       "686  Open Targets target identifier of the second m...   \n",
       "700  Ensembl identifier of the gene representing a ...   \n",
       "706  PubMed identifier of the publication hat refer...   \n",
       "736  Biosample identifier that maps to the biosampl...   \n",
       "\n",
       "                          foreign_key bioregistry_prefix             Example  \n",
       "14                               <NA>             pubmed     [PMID:14566559]  \n",
       "20                               <NA>                MGI         MGI:6140117  \n",
       "21                               <NA>             pubmed          [30949703]  \n",
       "23                               <NA>                 MP          MP:0005389  \n",
       "25                               <NA>                 MP          MP:0005343  \n",
       "29                               <NA>            ENSEMBL  ENSMUSG00000087651  \n",
       "30                               <NA>                MGI         MGI:1917034  \n",
       "48                               <NA>           reactome                 NaN  \n",
       "55                               <NA>            ENSEMBL     ENSG00000071243  \n",
       "57              biosample/biosampleId             UBERON                 NaN  \n",
       "83                               <NA>            ENSEMBL                 NaN  \n",
       "102                             go/id                 GO                 NaN  \n",
       "106                              <NA>            uniprot                 NaN  \n",
       "107                              <NA>                ECO                 NaN  \n",
       "110                              <NA>             pubmed                 NaN  \n",
       "114                              <NA>             pubmed                 NaN  \n",
       "131                              <NA>                 SL                 NaN  \n",
       "329                              <NA>                 go                 NaN  \n",
       "331                              <NA>                 so                 NaN  \n",
       "349                              <NA>             UBERON                 NaN  \n",
       "391                              <NA>          HANCESTRO                 NaN  \n",
       "409                              <NA>                 GO                 NaN  \n",
       "412                             so/id                 SO                 NaN  \n",
       "433             biosample/biosampleId             UBERON                 NaN  \n",
       "438                    disease_hpo/id                 HP                 NaN  \n",
       "441  mouse_phenotype/modelPhenotypeId                 MP                 NaN  \n",
       "444                  drug_molecule/id             CHEMBL                 NaN  \n",
       "456                             so/id                 SO                 NaN  \n",
       "466                       reactome/id           reactome                 NaN  \n",
       "496                              <NA>            ENSEMBL                 NaN  \n",
       "497                              <NA>                MGI                 NaN  \n",
       "505                             so/id                 SO                 NaN  \n",
       "541                             so/id                 SO                 NaN  \n",
       "548                         target/id            ENSEMBL                 NaN  \n",
       "559             biosample/biosampleId             UBERON                 NaN  \n",
       "596                         target/id            ENSEMBL                 NaN  \n",
       "599                              <NA>                 MI                 NaN  \n",
       "608                              <NA>                 MI                 NaN  \n",
       "609                              <NA>                 MI                 NaN  \n",
       "628                              <NA>                 MI                 NaN  \n",
       "632                         target/id            ENSEMBL                 NaN  \n",
       "633                              <NA>                 MI                 NaN  \n",
       "650                         target/id            ENSEMBL                 NaN  \n",
       "652                             so/id                 SO                 NaN  \n",
       "654                             so/id                 SO                 NaN  \n",
       "656                              <NA>            UNIPROT                 NaN  \n",
       "663                         target/id            ENSEMBL                 NaN  \n",
       "672                              <NA>            ENSEMBL                 NaN  \n",
       "683                         target/id            ENSEMBL                 NaN  \n",
       "686                         target/id            ENSEMBL                 NaN  \n",
       "700                         target/id            ENSEMBL                 NaN  \n",
       "706                   literature/pmid             PUBMED                 NaN  \n",
       "736             biosample/biosampleId             UBERON                 NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Columns with bioregistry prefix:\n",
    "curation_table.loc[curation_table.bioregistry_prefix.notna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>field_id</th>\n",
       "      <th>column_name</th>\n",
       "      <th>column_description</th>\n",
       "      <th>foreign_key</th>\n",
       "      <th>bioregistry_prefix</th>\n",
       "      <th>Example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/tissues/protein/reliability</td>\n",
       "      <td>reliability</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/tissues/protein/cell_type/reliability</td>\n",
       "      <td>reliability</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>expression</td>\n",
       "      <td>expression/tissues/protein/cell_type/level</td>\n",
       "      <td>level</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   dataset_name                                          field_id  \\\n",
       "67   expression            expression/tissues/protein/reliability   \n",
       "71   expression  expression/tissues/protein/cell_type/reliability   \n",
       "72   expression        expression/tissues/protein/cell_type/level   \n",
       "\n",
       "    column_name column_description foreign_key bioregistry_prefix Example  \n",
       "67  reliability               <NA>        <NA>                NaN     NaN  \n",
       "71  reliability               <NA>        <NA>                NaN     NaN  \n",
       "72        level               <NA>        <NA>                NaN     NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fields without description:\n",
    "curation_table.loc[curation_table.column_description.isna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Adding tags to distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>association_by_datasource_direct</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>association_by_datasource_indirect</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>association_by_datatype_direct</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>association_by_datatype_indirect</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>association_by_overall_indirect</td>\n",
       "      <td>[Target-Disease]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   id              tags\n",
       "0    association_by_datasource_direct  [Target-Disease]\n",
       "1  association_by_datasource_indirect  [Target-Disease]\n",
       "2      association_by_datatype_direct  [Target-Disease]\n",
       "3    association_by_datatype_indirect  [Target-Disease]\n",
       "4     association_by_overall_indirect  [Target-Disease]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "# Curation of all columns from all OpenTargets output datasets:\n",
    "curation = 'https://docs.google.com/spreadsheets/d/1JBu9HsRqwdGYMsYVoNAghmyEoU34S7fFnNIjHuzl0To/export?format=tsv'\n",
    "\n",
    "# Folder to save the resulting curation file:\n",
    "asset_folder = '../src/ot_croissant/assets/'\n",
    "\n",
    "# Reading table:\n",
    "curation_table = (\n",
    "    pd.read_csv(curation, sep='\\t')\n",
    "    .rename(columns={'Dataset': 'id', 'Tag': 'tag'})\n",
    "    .assign(\n",
    "        tags = lambda df: df.tag.str.split('|')\n",
    "    )\n",
    "    [['id', 'tags']]\n",
    ")\n",
    "\n",
    "curation_table.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collection of curated dataset:\n",
    "curation_json = []\n",
    "\n",
    "for _, row in (\n",
    "    pd.read_json(f'{asset_folder}/distribution.json', orient='records')\n",
    "    .merge(curation_table, on='id', how='left')\n",
    ").iterrows():\n",
    "    curation_json.append(\n",
    "        row.to_dict()\n",
    "    )\n",
    "\n",
    "\n",
    "# Saving the curation to a JSON file:\n",
    "with open(f'{asset_folder}/distribution.json', 'w') as f:\n",
    "    json.dump(curation_json, f, indent=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    pd.read_json(f'{asset_folder}/distribution.json', orient='records')\n",
    "    .merge(curation_table, on='id', how='left')\n",
    "    .to_json(f'{asset_folder}/distribution_v2.json', orient='table', indent=2, lines=False)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "25/04/29 11:41:10 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+-----------+--------+--------+---------------+----------------+-------+-------------+-----------+-----------+--------------------------+-------+----------+------------------+-------------+------------+-------+----------+--------+--------+----+--------------+-----------------------+-----------+\n",
      "| hm_variant_id|    hm_rsid|hm_chrom|  hm_pos|hm_other_allele|hm_effect_allele|hm_beta|hm_odds_ratio|hm_ci_lower|hm_ci_upper|hm_effect_allele_frequency|hm_code|chromosome|base_pair_location|effect_allele|other_allele|p_value|odds_ratio|ci_lower|ci_upper|beta|standard_error|effect_allele_frequency| variant_id|\n",
      "+--------------+-----------+--------+--------+---------------+----------------+-------+-------------+-----------+-----------+--------------------------+-------+----------+------------------+-------------+------------+-------+----------+--------+--------+----+--------------+-----------------------+-----------+\n",
      "|1_99534456_G_T| rs10875231|       1|99534456|              G|               T|     NA|           NA|         NA|         NA|                        NA|     10|         1|          99534456|            T|           G| 0.2697|        NA|      NA|      NA|  NA|            NA|                     NA| rs10875231|\n",
      "|1_99535271_C_T|  rs6678176|       1|99535271|              C|               T|     NA|           NA|         NA|         NA|                        NA|     10|         1|          99535271|            T|           C| 0.2028|        NA|      NA|      NA|  NA|            NA|                     NA|  rs6678176|\n",
      "|1_99535287_T_C| rs78286437|       1|99535287|              T|               C|     NA|           NA|         NA|         NA|                        NA|     11|         1|          99535287|            T|           C| 0.1592|        NA|      NA|      NA|  NA|            NA|                     NA| rs78286437|\n",
      "|1_99535645_G_T| rs76909621|       1|99535645|              G|               T|     NA|           NA|         NA|         NA|                        NA|     10|         1|          99535645|            T|           G|0.03791|        NA|      NA|      NA|  NA|            NA|                     NA| rs76909621|\n",
      "|1_99536934_C_T| rs78642210|       1|99536934|              C|               T|     NA|           NA|         NA|         NA|                        NA|     10|         1|          99536934|            T|           C| 0.1561|        NA|      NA|      NA|  NA|            NA|                     NA| rs78642210|\n",
      "|1_99537157_C_T| rs77140576|       1|99537157|              C|               T|     NA|           NA|         NA|         NA|                        NA|     10|         1|          99537157|            T|           C|0.03355|        NA|      NA|      NA|  NA|            NA|                     NA| rs77140576|\n",
      "|1_99537158_A_G|rs113470118|       1|99537158|              A|               G|     NA|           NA|         NA|         NA|                        NA|     11|         1|          99537158|            A|           G| 0.1419|        NA|      NA|      NA|  NA|            NA|                     NA|rs113470118|\n",
      "|1_99537326_T_G|  rs7545818|       1|99537326|              T|               G|     NA|           NA|         NA|         NA|                        NA|     11|         1|          99537326|            T|           G| 0.6851|        NA|      NA|      NA|  NA|            NA|                     NA|  rs7545818|\n",
      "|1_99537435_G_A| rs75635821|       1|99537435|              G|               A|     NA|           NA|         NA|         NA|                        NA|     10|         1|          99537435|            A|           G| 0.0336|        NA|      NA|      NA|  NA|            NA|                     NA| rs75635821|\n",
      "|1_99537648_T_G| rs78948828|       1|99537648|              T|               G|     NA|           NA|         NA|         NA|                        NA|     11|         1|          99537648|            T|           G| 0.1342|        NA|      NA|      NA|  NA|            NA|                     NA| rs78948828|\n",
      "| 1_9940342_T_A|  rs1237370|       1| 9940342|              T|               A|     NA|           NA|         NA|         NA|                        NA|      5|         1|           9940342|            A|           T| 0.3385|        NA|      NA|      NA|  NA|            NA|                     NA|  rs1237370|\n",
      "|1_99539170_G_A|  rs6682190|       1|99539170|              G|               A|     NA|           NA|         NA|         NA|                        NA|     10|         1|          99539170|            A|           G| 0.5989|        NA|      NA|      NA|  NA|            NA|                     NA|  rs6682190|\n",
      "|1_99539674_A_T|  rs6697069|       1|99539674|              A|               T|     NA|           NA|         NA|         NA|                        NA|      6|         1|          99539674|            A|           T|0.02979|        NA|      NA|      NA|  NA|            NA|                     NA|  rs6697069|\n",
      "|1_99539921_G_A| rs12069019|       1|99539921|              G|               A|     NA|           NA|         NA|         NA|                        NA|     10|         1|          99539921|            A|           G| 0.7602|        NA|      NA|      NA|  NA|            NA|                     NA| rs12069019|\n",
      "|1_99540561_G_A|  rs6686057|       1|99540561|              G|               A|     NA|           NA|         NA|         NA|                        NA|     10|         1|          99540561|            A|           G| 0.8952|        NA|      NA|      NA|  NA|            NA|                     NA|  rs6686057|\n",
      "|1_99541178_C_T| rs55725529|       1|99541178|              C|               T|     NA|           NA|         NA|         NA|                        NA|     10|         1|          99541178|            T|           C| 0.5069|        NA|      NA|      NA|  NA|            NA|                     NA| rs55725529|\n",
      "|1_99541898_C_T| rs12082355|       1|99541898|              C|               T|     NA|           NA|         NA|         NA|                        NA|     10|         1|          99541898|            T|           C|   0.75|        NA|      NA|      NA|  NA|            NA|                     NA| rs12082355|\n",
      "|1_99542185_A_G| rs12067343|       1|99542185|              A|               G|     NA|           NA|         NA|         NA|                        NA|     11|         1|          99542185|            A|           G| 0.5987|        NA|      NA|      NA|  NA|            NA|                     NA| rs12067343|\n",
      "|1_99543051_A_C| rs11166268|       1|99543051|              A|               C|     NA|           NA|         NA|         NA|                        NA|     11|         1|          99543051|            A|           C| 0.2104|        NA|      NA|      NA|  NA|            NA|                     NA| rs11166268|\n",
      "|1_99543431_C_A| rs11166269|       1|99543431|              C|               A|     NA|           NA|         NA|         NA|                        NA|     10|         1|          99543431|            A|           C| 0.6676|        NA|      NA|      NA|  NA|            NA|                     NA| rs11166269|\n",
      "+--------------+-----------+--------+--------+---------------+----------------+-------+-------------+-----------+-----------+--------------------------+-------+----------+------------------+-------------+------------+-------+----------+--------+--------+----+--------------+-----------------------+-----------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from pyspark.sql import SparkSession, functions as f, types as t\n",
    "\n",
    "\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "df = (\n",
    "    spark.read.csv('/Users/dsuveges/33414548-GCST90010801-EFO_0010226.h.tsv.gz', sep='\\t', header=True)\n",
    ")\n",
    "\n",
    "\n",
    "df.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "|"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
