{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "898780df",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql import DataFrame, SparkSession, functions as f\n",
    "import pandas as pd\n",
    "import json\n",
    "from urllib.error import URLError\n",
    "\n",
    "# https://docs.google.com/spreadsheets/d/132SKHMoaJePu4nTlBnQwfaz3dhfJiKmJUujfYkzXMdI/edit?gid=179018892#gid=179018892\n",
    "\n",
    "def get_evidence_datasets() -> pd.DataFrame:\n",
    "    # Evidence datasets:\n",
    "    spreadsheet_url = f'https://docs.google.com/spreadsheets/d/1G3zq-aJg3uKBz3VZUvGV8yYVrECtpQ9SJoGX6X3B-ms/export?format=csv'\n",
    "\n",
    "    try:\n",
    "        df = pd.read_csv(spreadsheet_url, sep=',')\n",
    "    except URLError:\n",
    "        df = pd.read_csv('dataset_name_mapping.csv')\n",
    "\n",
    "    # Extract table from spreadsheet:\n",
    "    return (\n",
    "        df\n",
    "        .loc[lambda df: df.output.notna()]\n",
    "        [['output', 'datasourceId']]\n",
    "    )\n",
    "\n",
    "def get_full_croissant(gid:str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    (\n",
    "        pd.read_csv(full_croissant_description_url,  sep=',')\n",
    "        .to_csv('full_croissant_description.csv', sep=',', index=False)\n",
    "    )\n",
    "    \"\"\"\n",
    "    # Evidence datasets:\n",
    "    full_croissant_description_url = f'https://docs.google.com/spreadsheets/d/132SKHMoaJePu4nTlBnQwfaz3dhfJiKmJUujfYkzXMdI/export?gid={gid}&format=csv'\n",
    "\n",
    "    try:\n",
    "        df = pd.read_csv(full_croissant_description_url, sep=',')\n",
    "    except URLError:\n",
    "        print('reading from backup: full_croissant_description.csv')\n",
    "        df = pd.read_csv('full_croissant_description.csv')\n",
    "\n",
    "    # Extract table from spreadsheet:\n",
    "    return df\n",
    "\n",
    "def get_full_evidence(gid: str) -> pd.DataFrame:\n",
    "    return (\n",
    "        get_full_croissant(gid)\n",
    "        .loc[lambda df: df.dataset_name == 'evidence']\n",
    "        .assign(\n",
    "            field_join = lambda x: x.field_id.str.extract(r'\\/(.+)$')\n",
    "        )\n",
    "        .drop(['dataset_name', 'Example', 'column_name', 'field_id'], axis=1)\n",
    "    )\n",
    "\n",
    "# 25.09:\n",
    "gid = '179018892'\n",
    "\n",
    "# Get a list of evidence datasets:\n",
    "dataset_list = get_evidence_datasets()\n",
    "\n",
    "# Reading a piece of dataset representing the schema of the entire dataset:\n",
    "spark = SparkSession.builder.getOrCreate()\n",
    "\n",
    "data_folder = '/Users/dsuveges/project_data/25.09/output'\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6ef322f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>output</th>\n",
       "      <th>datasourceId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>evidence_cancer_biomarkers</td>\n",
       "      <td>cancer_biomarkers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>evidence_cancer_gene_census</td>\n",
       "      <td>cancer_gene_census</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>evidence_chembl</td>\n",
       "      <td>chembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>evidence_clingen</td>\n",
       "      <td>clingen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>evidence_crispr</td>\n",
       "      <td>crispr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        output        datasourceId\n",
       "0   evidence_cancer_biomarkers   cancer_biomarkers\n",
       "1  evidence_cancer_gene_census  cancer_gene_census\n",
       "2              evidence_chembl              chembl\n",
       "3             evidence_clingen             clingen\n",
       "4              evidence_crispr              crispr"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_list.head() #to_csv('dataset_name_mapping.csv', sep=',', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a71c914",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "parent: evidence_cancer_biomarkers name: id\n",
      "parent: evidence_cancer_biomarkers name: targetFromSourceId\n",
      "parent: evidence_cancer_biomarkers name: diseaseFromSourceMappedId\n",
      "parent: evidence_cancer_biomarkers name: biomarkerName\n",
      "parent: evidence_cancer_biomarkers name: biomarkers\n",
      "dict\n",
      "struct\n",
      "parent: evidence_cancer_biomarkers/biomarkers name: geneExpression\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_cancer_biomarkers/biomarkers/geneExpression name: id\n",
      "parent: evidence_cancer_biomarkers/biomarkers/geneExpression name: name\n",
      "parent: evidence_cancer_biomarkers/biomarkers name: geneticVariation\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_cancer_biomarkers/biomarkers/geneticVariation name: functionalConsequenceId\n",
      "parent: evidence_cancer_biomarkers/biomarkers/geneticVariation name: id\n",
      "parent: evidence_cancer_biomarkers/biomarkers/geneticVariation name: name\n",
      "parent: evidence_cancer_biomarkers name: confidence\n",
      "parent: evidence_cancer_biomarkers name: datasourceId\n",
      "parent: evidence_cancer_biomarkers name: datatypeId\n",
      "parent: evidence_cancer_biomarkers name: diseaseFromSource\n",
      "parent: evidence_cancer_biomarkers name: drugFromSource\n",
      "parent: evidence_cancer_biomarkers name: drugId\n",
      "parent: evidence_cancer_biomarkers name: drugResponse\n",
      "parent: evidence_cancer_biomarkers name: literature\n",
      "dict\n",
      "parent: evidence_cancer_biomarkers name: urls\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_cancer_biomarkers/urls name: niceName\n",
      "parent: evidence_cancer_biomarkers/urls name: url\n",
      "parent: evidence_cancer_biomarkers name: qualityControls\n",
      "dict\n",
      "parent: evidence_cancer_biomarkers name: diseaseId\n",
      "parent: evidence_cancer_biomarkers name: targetId\n",
      "parent: evidence_cancer_biomarkers name: publicationDate\n",
      "parent: evidence_cancer_biomarkers name: evidenceDate\n",
      "parent: evidence_cancer_biomarkers name: score\n",
      "parent: evidence_cancer_gene_census name: targetId\n",
      "parent: evidence_cancer_gene_census name: id\n",
      "parent: evidence_cancer_gene_census name: targetFromSourceId\n",
      "parent: evidence_cancer_gene_census name: diseaseFromSourceMappedId\n",
      "parent: evidence_cancer_gene_census name: datasourceId\n",
      "parent: evidence_cancer_gene_census name: datatypeId\n",
      "parent: evidence_cancer_gene_census name: diseaseFromSourceId\n",
      "parent: evidence_cancer_gene_census name: literature\n",
      "dict\n",
      "parent: evidence_cancer_gene_census name: mutatedSamples\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_cancer_gene_census/mutatedSamples name: functionalConsequenceId\n",
      "parent: evidence_cancer_gene_census/mutatedSamples name: numberMutatedSamples\n",
      "parent: evidence_cancer_gene_census/mutatedSamples name: numberSamplesTested\n",
      "parent: evidence_cancer_gene_census/mutatedSamples name: numberSamplesWithMutationType\n",
      "parent: evidence_cancer_gene_census name: resourceScore\n",
      "parent: evidence_cancer_gene_census name: studyId\n",
      "parent: evidence_cancer_gene_census name: qualityControls\n",
      "dict\n",
      "parent: evidence_cancer_gene_census name: diseaseId\n",
      "parent: evidence_cancer_gene_census name: publicationDate\n",
      "parent: evidence_cancer_gene_census name: evidenceDate\n",
      "parent: evidence_cancer_gene_census name: score\n",
      "parent: evidence_cancer_gene_census name: directionOnTrait\n",
      "parent: evidence_cancer_gene_census name: directionOnTarget\n",
      "parent: evidence_chembl name: targetId\n",
      "parent: evidence_chembl name: drugId\n",
      "parent: evidence_chembl name: targetFromSourceId\n",
      "parent: evidence_chembl name: diseaseFromSourceMappedId\n",
      "parent: evidence_chembl name: clinicalPhase\n",
      "parent: evidence_chembl name: clinicalStatus\n",
      "parent: evidence_chembl name: cohortPhenotypes\n",
      "dict\n",
      "parent: evidence_chembl name: datasourceId\n",
      "parent: evidence_chembl name: datatypeId\n",
      "parent: evidence_chembl name: diseaseFromSource\n",
      "parent: evidence_chembl name: studyId\n",
      "parent: evidence_chembl name: studyStartDate\n",
      "parent: evidence_chembl name: studyStopReason\n",
      "parent: evidence_chembl name: studyStopReasonCategories\n",
      "dict\n",
      "parent: evidence_chembl name: targetFromSource\n",
      "parent: evidence_chembl name: urls\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_chembl/urls name: niceName\n",
      "parent: evidence_chembl/urls name: url\n",
      "parent: evidence_chembl name: qualityControls\n",
      "dict\n",
      "parent: evidence_chembl name: diseaseId\n",
      "parent: evidence_chembl name: id\n",
      "parent: evidence_chembl name: evidenceDate\n",
      "parent: evidence_chembl name: score\n",
      "parent: evidence_chembl name: directionOnTrait\n",
      "parent: evidence_chembl name: directionOnTarget\n",
      "parent: evidence_clingen name: targetFromSourceId\n",
      "parent: evidence_clingen name: diseaseFromSourceMappedId\n",
      "parent: evidence_clingen name: allelicRequirements\n",
      "dict\n",
      "parent: evidence_clingen name: confidence\n",
      "parent: evidence_clingen name: datasourceId\n",
      "parent: evidence_clingen name: datatypeId\n",
      "parent: evidence_clingen name: diseaseFromSource\n",
      "parent: evidence_clingen name: diseaseFromSourceId\n",
      "parent: evidence_clingen name: releaseDate\n",
      "parent: evidence_clingen name: studyId\n",
      "parent: evidence_clingen name: urls\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_clingen/urls name: url\n",
      "parent: evidence_clingen name: qualityControls\n",
      "dict\n",
      "parent: evidence_clingen name: diseaseId\n",
      "parent: evidence_clingen name: targetId\n",
      "parent: evidence_clingen name: id\n",
      "parent: evidence_clingen name: evidenceDate\n",
      "parent: evidence_clingen name: score\n",
      "parent: evidence_crispr name: id\n",
      "parent: evidence_crispr name: targetFromSourceId\n",
      "parent: evidence_crispr name: diseaseFromSourceMappedId\n",
      "parent: evidence_crispr name: datasourceId\n",
      "parent: evidence_crispr name: datatypeId\n",
      "parent: evidence_crispr name: diseaseCellLines\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_crispr/diseaseCellLines name: id\n",
      "parent: evidence_crispr/diseaseCellLines name: name\n",
      "parent: evidence_crispr/diseaseCellLines name: tissue\n",
      "parent: evidence_crispr/diseaseCellLines name: tissueId\n",
      "parent: evidence_crispr name: diseaseFromSource\n",
      "parent: evidence_crispr name: literature\n",
      "dict\n",
      "parent: evidence_crispr name: resourceScore\n",
      "parent: evidence_crispr name: targetFromSource\n",
      "parent: evidence_crispr name: qualityControls\n",
      "dict\n",
      "parent: evidence_crispr name: diseaseId\n",
      "parent: evidence_crispr name: targetId\n",
      "parent: evidence_crispr name: publicationDate\n",
      "parent: evidence_crispr name: evidenceDate\n",
      "parent: evidence_crispr name: score\n",
      "parent: evidence_crispr_screen name: id\n",
      "parent: evidence_crispr_screen name: targetFromSourceId\n",
      "parent: evidence_crispr_screen name: diseaseFromSourceMappedId\n",
      "parent: evidence_crispr_screen name: cellType\n",
      "parent: evidence_crispr_screen name: contrast\n",
      "parent: evidence_crispr_screen name: crisprScreenLibrary\n",
      "parent: evidence_crispr_screen name: datasourceId\n",
      "parent: evidence_crispr_screen name: datatypeId\n",
      "parent: evidence_crispr_screen name: diseaseFromSource\n",
      "parent: evidence_crispr_screen name: geneticBackground\n",
      "parent: evidence_crispr_screen name: literature\n",
      "dict\n",
      "parent: evidence_crispr_screen name: log2FoldChangeValue\n",
      "parent: evidence_crispr_screen name: projectId\n",
      "parent: evidence_crispr_screen name: resourceScore\n",
      "parent: evidence_crispr_screen name: statisticalTestTail\n",
      "parent: evidence_crispr_screen name: studyId\n",
      "parent: evidence_crispr_screen name: studyOverview\n",
      "parent: evidence_crispr_screen name: qualityControls\n",
      "dict\n",
      "parent: evidence_crispr_screen name: diseaseId\n",
      "parent: evidence_crispr_screen name: targetId\n",
      "parent: evidence_crispr_screen name: publicationDate\n",
      "parent: evidence_crispr_screen name: evidenceDate\n",
      "parent: evidence_crispr_screen name: score\n",
      "parent: evidence_encore name: targetFromSourceId\n",
      "parent: evidence_encore name: diseaseFromSourceMappedId\n",
      "parent: evidence_encore name: biomarkerList\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_encore/biomarkerList name: description\n",
      "parent: evidence_encore/biomarkerList name: name\n",
      "parent: evidence_encore name: datasourceId\n",
      "parent: evidence_encore name: datatypeId\n",
      "parent: evidence_encore name: diseaseCellLines\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_encore/diseaseCellLines name: id\n",
      "parent: evidence_encore/diseaseCellLines name: name\n",
      "parent: evidence_encore/diseaseCellLines name: tissue\n",
      "parent: evidence_encore/diseaseCellLines name: tissueId\n",
      "parent: evidence_encore name: diseaseFromSource\n",
      "parent: evidence_encore name: geneInteractionType\n",
      "parent: evidence_encore name: geneticInteractionPValue\n",
      "parent: evidence_encore name: geneticInteractionScore\n",
      "parent: evidence_encore name: interactingTargetFromSourceId\n",
      "parent: evidence_encore name: interactingTargetRole\n",
      "parent: evidence_encore name: phenotypicConsequenceFDR\n",
      "parent: evidence_encore name: phenotypicConsequenceLogFoldChange\n",
      "parent: evidence_encore name: phenotypicConsequencePValue\n",
      "parent: evidence_encore name: projectDescription\n",
      "parent: evidence_encore name: projectId\n",
      "parent: evidence_encore name: releaseDate\n",
      "parent: evidence_encore name: releaseVersion\n",
      "parent: evidence_encore name: statisticalMethod\n",
      "parent: evidence_encore name: targetRole\n",
      "parent: evidence_encore name: qualityControls\n",
      "dict\n",
      "parent: evidence_encore name: diseaseId\n",
      "parent: evidence_encore name: targetId\n",
      "parent: evidence_encore name: id\n",
      "parent: evidence_encore name: evidenceDate\n",
      "parent: evidence_encore name: score\n",
      "parent: evidence_europepmc name: id\n",
      "parent: evidence_europepmc name: targetFromSourceId\n",
      "parent: evidence_europepmc name: diseaseFromSourceMappedId\n",
      "parent: evidence_europepmc name: datasourceId\n",
      "parent: evidence_europepmc name: datatypeId\n",
      "parent: evidence_europepmc name: literature\n",
      "dict\n",
      "parent: evidence_europepmc name: pmcIds\n",
      "dict\n",
      "parent: evidence_europepmc name: publicationYear\n",
      "parent: evidence_europepmc name: resourceScore\n",
      "parent: evidence_europepmc name: textMiningSentences\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_europepmc/textMiningSentences name: dEnd\n",
      "parent: evidence_europepmc/textMiningSentences name: dStart\n",
      "parent: evidence_europepmc/textMiningSentences name: section\n",
      "parent: evidence_europepmc/textMiningSentences name: tEnd\n",
      "parent: evidence_europepmc/textMiningSentences name: tStart\n",
      "parent: evidence_europepmc/textMiningSentences name: text\n",
      "parent: evidence_europepmc name: qualityControls\n",
      "dict\n",
      "parent: evidence_europepmc name: diseaseId\n",
      "parent: evidence_europepmc name: targetId\n",
      "parent: evidence_europepmc name: publicationDate\n",
      "parent: evidence_europepmc name: evidenceDate\n",
      "parent: evidence_europepmc name: score\n",
      "parent: evidence_eva name: targetId\n",
      "parent: evidence_eva name: id\n",
      "parent: evidence_eva name: targetFromSourceId\n",
      "parent: evidence_eva name: diseaseFromSourceMappedId\n",
      "parent: evidence_eva name: alleleOrigins\n",
      "dict\n",
      "parent: evidence_eva name: allelicRequirements\n",
      "dict\n",
      "parent: evidence_eva name: clinicalSignificances\n",
      "dict\n",
      "parent: evidence_eva name: cohortPhenotypes\n",
      "dict\n",
      "parent: evidence_eva name: confidence\n",
      "parent: evidence_eva name: datasourceId\n",
      "parent: evidence_eva name: datatypeId\n",
      "parent: evidence_eva name: diseaseFromSource\n",
      "parent: evidence_eva name: diseaseFromSourceId\n",
      "parent: evidence_eva name: literature\n",
      "dict\n",
      "parent: evidence_eva name: releaseDate\n",
      "parent: evidence_eva name: studyId\n",
      "parent: evidence_eva name: variantFromSourceId\n",
      "parent: evidence_eva name: variantFunctionalConsequenceId\n",
      "parent: evidence_eva name: variantHgvsId\n",
      "parent: evidence_eva name: variantId\n",
      "parent: evidence_eva name: variantRsId\n",
      "parent: evidence_eva name: qualityControls\n",
      "dict\n",
      "parent: evidence_eva name: diseaseId\n",
      "parent: evidence_eva name: publicationDate\n",
      "parent: evidence_eva name: evidenceDate\n",
      "parent: evidence_eva name: score\n",
      "parent: evidence_eva name: directionOnTrait\n",
      "parent: evidence_eva name: directionOnTarget\n",
      "parent: evidence_eva_somatic name: targetId\n",
      "parent: evidence_eva_somatic name: id\n",
      "parent: evidence_eva_somatic name: targetFromSourceId\n",
      "parent: evidence_eva_somatic name: diseaseFromSourceMappedId\n",
      "parent: evidence_eva_somatic name: alleleOrigins\n",
      "dict\n",
      "parent: evidence_eva_somatic name: allelicRequirements\n",
      "dict\n",
      "parent: evidence_eva_somatic name: clinicalSignificances\n",
      "dict\n",
      "parent: evidence_eva_somatic name: cohortPhenotypes\n",
      "dict\n",
      "parent: evidence_eva_somatic name: confidence\n",
      "parent: evidence_eva_somatic name: datasourceId\n",
      "parent: evidence_eva_somatic name: datatypeId\n",
      "parent: evidence_eva_somatic name: diseaseFromSource\n",
      "parent: evidence_eva_somatic name: diseaseFromSourceId\n",
      "parent: evidence_eva_somatic name: literature\n",
      "dict\n",
      "parent: evidence_eva_somatic name: releaseDate\n",
      "parent: evidence_eva_somatic name: studyId\n",
      "parent: evidence_eva_somatic name: variantFromSourceId\n",
      "parent: evidence_eva_somatic name: variantFunctionalConsequenceId\n",
      "parent: evidence_eva_somatic name: variantHgvsId\n",
      "parent: evidence_eva_somatic name: variantId\n",
      "parent: evidence_eva_somatic name: variantRsId\n",
      "parent: evidence_eva_somatic name: qualityControls\n",
      "dict\n",
      "parent: evidence_eva_somatic name: diseaseId\n",
      "parent: evidence_eva_somatic name: publicationDate\n",
      "parent: evidence_eva_somatic name: evidenceDate\n",
      "parent: evidence_eva_somatic name: score\n",
      "parent: evidence_eva_somatic name: directionOnTrait\n",
      "parent: evidence_eva_somatic name: directionOnTarget\n",
      "parent: evidence_expression_atlas name: id\n",
      "parent: evidence_expression_atlas name: targetFromSourceId\n",
      "parent: evidence_expression_atlas name: diseaseFromSourceMappedId\n",
      "parent: evidence_expression_atlas name: biosamplesFromSource\n",
      "dict\n",
      "parent: evidence_expression_atlas name: confidence\n",
      "parent: evidence_expression_atlas name: contrast\n",
      "parent: evidence_expression_atlas name: datasourceId\n",
      "parent: evidence_expression_atlas name: datatypeId\n",
      "parent: evidence_expression_atlas name: literature\n",
      "dict\n",
      "parent: evidence_expression_atlas name: log2FoldChangePercentileRank\n",
      "parent: evidence_expression_atlas name: log2FoldChangeValue\n",
      "parent: evidence_expression_atlas name: resourceScore\n",
      "parent: evidence_expression_atlas name: studyId\n",
      "parent: evidence_expression_atlas name: studyOverview\n",
      "parent: evidence_expression_atlas name: qualityControls\n",
      "dict\n",
      "parent: evidence_expression_atlas name: diseaseId\n",
      "parent: evidence_expression_atlas name: targetId\n",
      "parent: evidence_expression_atlas name: publicationDate\n",
      "parent: evidence_expression_atlas name: evidenceDate\n",
      "parent: evidence_expression_atlas name: score\n",
      "parent: evidence_gene_burden name: targetId\n",
      "parent: evidence_gene_burden name: id\n",
      "parent: evidence_gene_burden name: targetFromSourceId\n",
      "parent: evidence_gene_burden name: diseaseFromSourceMappedId\n",
      "parent: evidence_gene_burden name: allelicRequirements\n",
      "dict\n",
      "parent: evidence_gene_burden name: ancestry\n",
      "parent: evidence_gene_burden name: ancestryId\n",
      "parent: evidence_gene_burden name: beta\n",
      "parent: evidence_gene_burden name: betaConfidenceIntervalLower\n",
      "parent: evidence_gene_burden name: betaConfidenceIntervalUpper\n",
      "parent: evidence_gene_burden name: cohortId\n",
      "parent: evidence_gene_burden name: datasourceId\n",
      "parent: evidence_gene_burden name: datatypeId\n",
      "parent: evidence_gene_burden name: diseaseFromSource\n",
      "parent: evidence_gene_burden name: diseaseFromSourceId\n",
      "parent: evidence_gene_burden name: literature\n",
      "dict\n",
      "parent: evidence_gene_burden name: oddsRatio\n",
      "parent: evidence_gene_burden name: oddsRatioConfidenceIntervalLower\n",
      "parent: evidence_gene_burden name: oddsRatioConfidenceIntervalUpper\n",
      "parent: evidence_gene_burden name: pValueExponent\n",
      "parent: evidence_gene_burden name: pValueMantissa\n",
      "parent: evidence_gene_burden name: projectId\n",
      "parent: evidence_gene_burden name: releaseVersion\n",
      "parent: evidence_gene_burden name: resourceScore\n",
      "parent: evidence_gene_burden name: sex\n",
      "dict\n",
      "parent: evidence_gene_burden name: statisticalMethod\n",
      "parent: evidence_gene_burden name: statisticalMethodOverview\n",
      "parent: evidence_gene_burden name: studyCases\n",
      "parent: evidence_gene_burden name: studyCasesWithQualifyingVariants\n",
      "parent: evidence_gene_burden name: studySampleSize\n",
      "parent: evidence_gene_burden name: urls\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_gene_burden/urls name: niceName\n",
      "parent: evidence_gene_burden/urls name: url\n",
      "parent: evidence_gene_burden name: qualityControls\n",
      "dict\n",
      "parent: evidence_gene_burden name: diseaseId\n",
      "parent: evidence_gene_burden name: publicationDate\n",
      "parent: evidence_gene_burden name: evidenceDate\n",
      "parent: evidence_gene_burden name: score\n",
      "parent: evidence_gene_burden name: directionOnTrait\n",
      "parent: evidence_gene_burden name: directionOnTarget\n",
      "parent: evidence_gene2phenotype name: targetId\n",
      "parent: evidence_gene2phenotype name: id\n",
      "parent: evidence_gene2phenotype name: targetFromSourceId\n",
      "parent: evidence_gene2phenotype name: diseaseFromSourceMappedId\n",
      "parent: evidence_gene2phenotype name: allelicRequirements\n",
      "dict\n",
      "parent: evidence_gene2phenotype name: confidence\n",
      "parent: evidence_gene2phenotype name: datasourceId\n",
      "parent: evidence_gene2phenotype name: datatypeId\n",
      "parent: evidence_gene2phenotype name: diseaseFromSource\n",
      "parent: evidence_gene2phenotype name: diseaseFromSourceId\n",
      "parent: evidence_gene2phenotype name: literature\n",
      "dict\n",
      "parent: evidence_gene2phenotype name: studyId\n",
      "parent: evidence_gene2phenotype name: variantFunctionalConsequenceId\n",
      "parent: evidence_gene2phenotype name: qualityControls\n",
      "dict\n",
      "parent: evidence_gene2phenotype name: diseaseId\n",
      "parent: evidence_gene2phenotype name: publicationDate\n",
      "parent: evidence_gene2phenotype name: evidenceDate\n",
      "parent: evidence_gene2phenotype name: score\n",
      "parent: evidence_gene2phenotype name: directionOnTrait\n",
      "parent: evidence_gene2phenotype name: directionOnTarget\n",
      "parent: evidence_genomics_england name: id\n",
      "parent: evidence_genomics_england name: targetFromSourceId\n",
      "parent: evidence_genomics_england name: diseaseFromSourceMappedId\n",
      "parent: evidence_genomics_england name: allelicRequirements\n",
      "dict\n",
      "parent: evidence_genomics_england name: cohortPhenotypes\n",
      "dict\n",
      "parent: evidence_genomics_england name: confidence\n",
      "parent: evidence_genomics_england name: datasourceId\n",
      "parent: evidence_genomics_england name: datatypeId\n",
      "parent: evidence_genomics_england name: diseaseFromSource\n",
      "parent: evidence_genomics_england name: diseaseFromSourceId\n",
      "parent: evidence_genomics_england name: literature\n",
      "dict\n",
      "parent: evidence_genomics_england name: studyId\n",
      "parent: evidence_genomics_england name: studyOverview\n",
      "parent: evidence_genomics_england name: qualityControls\n",
      "dict\n",
      "parent: evidence_genomics_england name: diseaseId\n",
      "parent: evidence_genomics_england name: targetId\n",
      "parent: evidence_genomics_england name: publicationDate\n",
      "parent: evidence_genomics_england name: evidenceDate\n",
      "parent: evidence_genomics_england name: score\n",
      "parent: evidence_gwas_credible_sets name: id\n",
      "parent: evidence_gwas_credible_sets name: targetFromSourceId\n",
      "parent: evidence_gwas_credible_sets name: diseaseFromSourceMappedId\n",
      "parent: evidence_gwas_credible_sets name: datasourceId\n",
      "parent: evidence_gwas_credible_sets name: datatypeId\n",
      "parent: evidence_gwas_credible_sets name: literature\n",
      "dict\n",
      "parent: evidence_gwas_credible_sets name: resourceScore\n",
      "parent: evidence_gwas_credible_sets name: studyLocusId\n",
      "parent: evidence_gwas_credible_sets name: qualityControls\n",
      "dict\n",
      "parent: evidence_gwas_credible_sets name: diseaseId\n",
      "parent: evidence_gwas_credible_sets name: targetId\n",
      "parent: evidence_gwas_credible_sets name: publicationDate\n",
      "parent: evidence_gwas_credible_sets name: evidenceDate\n",
      "parent: evidence_gwas_credible_sets name: score\n",
      "parent: evidence_impc name: targetId\n",
      "parent: evidence_impc name: id\n",
      "parent: evidence_impc name: targetFromSourceId\n",
      "parent: evidence_impc name: diseaseFromSourceMappedId\n",
      "parent: evidence_impc name: biologicalModelAllelicComposition\n",
      "parent: evidence_impc name: biologicalModelGeneticBackground\n",
      "parent: evidence_impc name: biologicalModelId\n",
      "parent: evidence_impc name: datasourceId\n",
      "parent: evidence_impc name: datatypeId\n",
      "parent: evidence_impc name: diseaseFromSource\n",
      "parent: evidence_impc name: diseaseFromSourceId\n",
      "parent: evidence_impc name: diseaseModelAssociatedHumanPhenotypes\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_impc/diseaseModelAssociatedHumanPhenotypes name: id\n",
      "parent: evidence_impc/diseaseModelAssociatedHumanPhenotypes name: label\n",
      "parent: evidence_impc name: diseaseModelAssociatedModelPhenotypes\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_impc/diseaseModelAssociatedModelPhenotypes name: id\n",
      "parent: evidence_impc/diseaseModelAssociatedModelPhenotypes name: label\n",
      "parent: evidence_impc name: literature\n",
      "dict\n",
      "parent: evidence_impc name: resourceScore\n",
      "parent: evidence_impc name: targetInModel\n",
      "parent: evidence_impc name: targetInModelEnsemblId\n",
      "parent: evidence_impc name: targetInModelMgiId\n",
      "parent: evidence_impc name: qualityControls\n",
      "dict\n",
      "parent: evidence_impc name: diseaseId\n",
      "parent: evidence_impc name: publicationDate\n",
      "parent: evidence_impc name: evidenceDate\n",
      "parent: evidence_impc name: score\n",
      "parent: evidence_impc name: directionOnTrait\n",
      "parent: evidence_impc name: directionOnTarget\n",
      "parent: evidence_intogen name: targetId\n",
      "parent: evidence_intogen name: id\n",
      "parent: evidence_intogen name: targetFromSourceId\n",
      "parent: evidence_intogen name: diseaseFromSourceMappedId\n",
      "parent: evidence_intogen name: cohortDescription\n",
      "parent: evidence_intogen name: cohortId\n",
      "parent: evidence_intogen name: cohortShortName\n",
      "parent: evidence_intogen name: datasourceId\n",
      "parent: evidence_intogen name: datatypeId\n",
      "parent: evidence_intogen name: diseaseFromSource\n",
      "parent: evidence_intogen name: literature\n",
      "dict\n",
      "parent: evidence_intogen name: mutatedSamples\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_intogen/mutatedSamples name: functionalConsequenceId\n",
      "parent: evidence_intogen/mutatedSamples name: numberMutatedSamples\n",
      "parent: evidence_intogen/mutatedSamples name: numberSamplesTested\n",
      "parent: evidence_intogen name: resourceScore\n",
      "parent: evidence_intogen name: significantDriverMethods\n",
      "dict\n",
      "parent: evidence_intogen name: qualityControls\n",
      "dict\n",
      "parent: evidence_intogen name: diseaseId\n",
      "parent: evidence_intogen name: publicationDate\n",
      "parent: evidence_intogen name: evidenceDate\n",
      "parent: evidence_intogen name: score\n",
      "parent: evidence_intogen name: directionOnTrait\n",
      "parent: evidence_intogen name: directionOnTarget\n",
      "parent: evidence_orphanet name: targetId\n",
      "parent: evidence_orphanet name: id\n",
      "parent: evidence_orphanet name: targetFromSourceId\n",
      "parent: evidence_orphanet name: diseaseFromSourceMappedId\n",
      "parent: evidence_orphanet name: alleleOrigins\n",
      "dict\n",
      "parent: evidence_orphanet name: confidence\n",
      "parent: evidence_orphanet name: datasourceId\n",
      "parent: evidence_orphanet name: datatypeId\n",
      "parent: evidence_orphanet name: diseaseFromSource\n",
      "parent: evidence_orphanet name: diseaseFromSourceId\n",
      "parent: evidence_orphanet name: literature\n",
      "dict\n",
      "parent: evidence_orphanet name: targetFromSource\n",
      "parent: evidence_orphanet name: variantFunctionalConsequenceId\n",
      "parent: evidence_orphanet name: qualityControls\n",
      "dict\n",
      "parent: evidence_orphanet name: diseaseId\n",
      "parent: evidence_orphanet name: publicationDate\n",
      "parent: evidence_orphanet name: evidenceDate\n",
      "parent: evidence_orphanet name: score\n",
      "parent: evidence_orphanet name: directionOnTrait\n",
      "parent: evidence_orphanet name: directionOnTarget\n",
      "parent: evidence_ot_crispr name: targetFromSourceId\n",
      "parent: evidence_ot_crispr name: diseaseFromSourceMappedId\n",
      "parent: evidence_ot_crispr name: cellLineBackground\n",
      "parent: evidence_ot_crispr name: cellType\n",
      "parent: evidence_ot_crispr name: contrast\n",
      "parent: evidence_ot_crispr name: crisprScreenLibrary\n",
      "parent: evidence_ot_crispr name: datasourceId\n",
      "parent: evidence_ot_crispr name: datatypeId\n",
      "parent: evidence_ot_crispr name: log2FoldChangeValue\n",
      "parent: evidence_ot_crispr name: projectDescription\n",
      "parent: evidence_ot_crispr name: projectId\n",
      "parent: evidence_ot_crispr name: releaseDate\n",
      "parent: evidence_ot_crispr name: releaseVersion\n",
      "parent: evidence_ot_crispr name: resourceScore\n",
      "parent: evidence_ot_crispr name: statisticalTestTail\n",
      "parent: evidence_ot_crispr name: studyId\n",
      "parent: evidence_ot_crispr name: studyOverview\n",
      "parent: evidence_ot_crispr name: qualityControls\n",
      "dict\n",
      "parent: evidence_ot_crispr name: diseaseId\n",
      "parent: evidence_ot_crispr name: targetId\n",
      "parent: evidence_ot_crispr name: id\n",
      "parent: evidence_ot_crispr name: evidenceDate\n",
      "parent: evidence_ot_crispr name: score\n",
      "parent: evidence_ot_crispr_validation name: targetFromSourceId\n",
      "parent: evidence_ot_crispr_validation name: diseaseFromSourceMappedId\n",
      "parent: evidence_ot_crispr_validation name: assays\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_ot_crispr_validation/assays name: description\n",
      "parent: evidence_ot_crispr_validation/assays name: isHit\n",
      "parent: evidence_ot_crispr_validation/assays name: shortName\n",
      "parent: evidence_ot_crispr_validation name: assessments\n",
      "dict\n",
      "parent: evidence_ot_crispr_validation name: biomarkerList\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_ot_crispr_validation/biomarkerList name: description\n",
      "parent: evidence_ot_crispr_validation/biomarkerList name: name\n",
      "parent: evidence_ot_crispr_validation name: datasourceId\n",
      "parent: evidence_ot_crispr_validation name: datatypeId\n",
      "parent: evidence_ot_crispr_validation name: diseaseCellLines\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_ot_crispr_validation/diseaseCellLines name: id\n",
      "parent: evidence_ot_crispr_validation/diseaseCellLines name: name\n",
      "parent: evidence_ot_crispr_validation/diseaseCellLines name: tissue\n",
      "parent: evidence_ot_crispr_validation/diseaseCellLines name: tissueId\n",
      "parent: evidence_ot_crispr_validation name: diseaseFromSource\n",
      "parent: evidence_ot_crispr_validation name: primaryProjectHit\n",
      "parent: evidence_ot_crispr_validation name: primaryProjectId\n",
      "parent: evidence_ot_crispr_validation name: projectId\n",
      "parent: evidence_ot_crispr_validation name: releaseDate\n",
      "parent: evidence_ot_crispr_validation name: releaseVersion\n",
      "parent: evidence_ot_crispr_validation name: resourceScore\n",
      "parent: evidence_ot_crispr_validation name: studyOverview\n",
      "parent: evidence_ot_crispr_validation name: qualityControls\n",
      "dict\n",
      "parent: evidence_ot_crispr_validation name: diseaseId\n",
      "parent: evidence_ot_crispr_validation name: targetId\n",
      "parent: evidence_ot_crispr_validation name: id\n",
      "parent: evidence_ot_crispr_validation name: evidenceDate\n",
      "parent: evidence_ot_crispr_validation name: score\n",
      "parent: evidence_reactome name: id\n",
      "parent: evidence_reactome name: targetFromSourceId\n",
      "parent: evidence_reactome name: diseaseFromSourceMappedId\n",
      "parent: evidence_reactome name: datasourceId\n",
      "parent: evidence_reactome name: datatypeId\n",
      "parent: evidence_reactome name: diseaseFromSource\n",
      "parent: evidence_reactome name: diseaseFromSourceId\n",
      "parent: evidence_reactome name: literature\n",
      "dict\n",
      "parent: evidence_reactome name: pathways\n",
      "dict\n",
      "array of struct\n",
      "parent: evidence_reactome/pathways name: id\n",
      "parent: evidence_reactome/pathways name: name\n",
      "parent: evidence_reactome name: reactionId\n",
      "parent: evidence_reactome name: reactionName\n",
      "parent: evidence_reactome name: targetModulation\n",
      "parent: evidence_reactome name: variantAminoacidDescriptions\n",
      "dict\n",
      "parent: evidence_reactome name: qualityControls\n",
      "dict\n",
      "parent: evidence_reactome name: diseaseId\n",
      "parent: evidence_reactome name: targetId\n",
      "parent: evidence_reactome name: publicationDate\n",
      "parent: evidence_reactome name: evidenceDate\n",
      "parent: evidence_reactome name: score\n",
      "parent: evidence_uniprot_literature name: id\n",
      "parent: evidence_uniprot_literature name: targetFromSourceId\n",
      "parent: evidence_uniprot_literature name: diseaseFromSourceMappedId\n",
      "parent: evidence_uniprot_literature name: confidence\n",
      "parent: evidence_uniprot_literature name: datasourceId\n",
      "parent: evidence_uniprot_literature name: datatypeId\n",
      "parent: evidence_uniprot_literature name: diseaseFromSource\n",
      "parent: evidence_uniprot_literature name: diseaseFromSourceId\n",
      "parent: evidence_uniprot_literature name: literature\n",
      "dict\n",
      "parent: evidence_uniprot_literature name: targetModulation\n",
      "parent: evidence_uniprot_literature name: qualityControls\n",
      "dict\n",
      "parent: evidence_uniprot_literature name: diseaseId\n",
      "parent: evidence_uniprot_literature name: targetId\n",
      "parent: evidence_uniprot_literature name: publicationDate\n",
      "parent: evidence_uniprot_literature name: evidenceDate\n",
      "parent: evidence_uniprot_literature name: score\n",
      "parent: evidence_uniprot_variants name: id\n",
      "parent: evidence_uniprot_variants name: targetFromSourceId\n",
      "parent: evidence_uniprot_variants name: diseaseFromSourceMappedId\n",
      "parent: evidence_uniprot_variants name: confidence\n",
      "parent: evidence_uniprot_variants name: datasourceId\n",
      "parent: evidence_uniprot_variants name: datatypeId\n",
      "parent: evidence_uniprot_variants name: diseaseFromSource\n",
      "parent: evidence_uniprot_variants name: diseaseFromSourceId\n",
      "parent: evidence_uniprot_variants name: literature\n",
      "dict\n",
      "parent: evidence_uniprot_variants name: targetModulation\n",
      "parent: evidence_uniprot_variants name: variantId\n",
      "parent: evidence_uniprot_variants name: variantRsId\n",
      "parent: evidence_uniprot_variants name: qualityControls\n",
      "dict\n",
      "parent: evidence_uniprot_variants name: diseaseId\n",
      "parent: evidence_uniprot_variants name: targetId\n",
      "parent: evidence_uniprot_variants name: publicationDate\n",
      "parent: evidence_uniprot_variants name: evidenceDate\n",
      "parent: evidence_uniprot_variants name: score\n"
     ]
    }
   ],
   "source": [
    "# Processing a field in the spark schema:\n",
    "# Processing a field in the spark schema:\n",
    "def process_fields(\n",
    "    fields: dict[str, str], \n",
    "    parent: str, \n",
    "    dataset_name: str,\n",
    "    schema: list[dict[str,str]] | None = None,\n",
    ") -> list[dict[str, str]]:\n",
    "    if schema is None:\n",
    "        schema = []\n",
    "\n",
    "    # Iterating over all fields of the schema:\n",
    "    for field in fields.get('fields'):\n",
    "\n",
    "        name = field.get('name')\n",
    "        # Recurse for nested structs:\n",
    "        if isinstance(field['type'], dict):\n",
    "            if field['type']['type'] == 'struct':\n",
    "                # Call self:\n",
    "                schema = schema + process_fields(\n",
    "                    field['type'], \n",
    "                    f'{parent}/{name}',\n",
    "                    dataset_name,\n",
    "                    [],\n",
    "                )\n",
    "            elif isinstance(field['type']['elementType'], dict):\n",
    "                # Call self:\n",
    "                schema = schema + process_fields(\n",
    "                    field[\"type\"]['elementType'], \n",
    "                    f'{parent}/{name}',\n",
    "                    dataset_name,\n",
    "                    [],\n",
    "                )\n",
    "        # Capture atomic column types:\n",
    "        schema.append(\n",
    "            {\n",
    "                'dataset_name': dataset_name,\n",
    "                'field_id': f'{parent}/{name}',\n",
    "                'column_name': name,\n",
    "            }\n",
    "        )\n",
    "    # Return captured schema:\n",
    "    return schema\n",
    "\n",
    "\n",
    "def generate_schema_representation(\n",
    "    output: str,\n",
    "    datasourceId: str\n",
    ") -> pd.DataFrame:\n",
    "\n",
    "    # Read dataset:\n",
    "    try:\n",
    "        df = spark.read.parquet(f'{data_folder}/{output}')\n",
    "    except:\n",
    "        print(f'missing dataset: {data_folder}/{output}')\n",
    "        return None\n",
    "\n",
    "    # Get its schema:\n",
    "    data = json.loads(df.schema.json())\n",
    "\n",
    "    # We iterate over the schema and collect field names that we will be able to annotate on the spreadsheet:\n",
    "    schema = process_fields(\n",
    "        data, \n",
    "        output,\n",
    "        output\n",
    "    )\n",
    "\n",
    "    # Process and return schema representation:\n",
    "    return (\n",
    "        pd.DataFrame(schema)\n",
    "        .assign(\n",
    "            field_join = lambda x: x.field_id.str.extract(r'\\/(.+)$')\n",
    "        )\n",
    "    )\n",
    "\n",
    "\n",
    "\n",
    "# generate_schema_representation(**dataset)    \n",
    "new_evidence_collated = pd.concat([\n",
    "    generate_schema_representation(**row)\n",
    "    for _, row in \n",
    "    dataset_list.iterrows()\n",
    "])\n",
    "\n",
    "# print(len(new_evidence_collated))\n",
    "# new_evidence_collated.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "420a502c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>column_description</th>\n",
       "      <th>foreign_key</th>\n",
       "      <th>bioregistry_prefix</th>\n",
       "      <th>field_join</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>Identifer of the evidence source</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>datasourceId</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>Open Targets target identifier</td>\n",
       "      <td>target/id</td>\n",
       "      <td>NaN</td>\n",
       "      <td>targetId</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>Origin of the variant allele</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>alleleOrigins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>Inheritance patterns</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>allelicRequirements</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>Genetic origin of a population</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ancestry</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   column_description foreign_key bioregistry_prefix  \\\n",
       "202  Identifer of the evidence source         NaN                NaN   \n",
       "203    Open Targets target identifier   target/id                NaN   \n",
       "204      Origin of the variant allele         NaN                NaN   \n",
       "205              Inheritance patterns         NaN                NaN   \n",
       "206    Genetic origin of a population         NaN                NaN   \n",
       "\n",
       "              field_join  \n",
       "202         datasourceId  \n",
       "203             targetId  \n",
       "204        alleleOrigins  \n",
       "205  allelicRequirements  \n",
       "206             ancestry  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "old_evidence = get_full_evidence(gid)\n",
    "old_evidence.head() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "462a0a4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>field_id</th>\n",
       "      <th>column_name</th>\n",
       "      <th>field_join</th>\n",
       "      <th>column_description</th>\n",
       "      <th>foreign_key</th>\n",
       "      <th>bioregistry_prefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [dataset_name, field_id, column_name, field_join, column_description, foreign_key, bioregistry_prefix]\n",
       "Index: []"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "updated_evidence_table = (\n",
    "    new_evidence_collated\n",
    "    .merge(\n",
    "        old_evidence, on='field_join', how='left'\n",
    "    )\n",
    "    # Updating missing descriptions:\n",
    "    .assign(\n",
    "        column_description = lambda df: df.apply(\n",
    "            lambda row: \n",
    "              \"Evidence quality flags\" if row['column_name'] == 'qualityControls' else \n",
    "              \"Effect of direction on target\" if row['column_name'] == 'directionOnTarget' else\n",
    "              row['column_description'],\n",
    "            axis=1\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "\n",
    "updated_evidence_table.loc[lambda df: df.column_description.isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "56b7db65",
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    pd.concat([\n",
    "        updated_evidence_table,\n",
    "        get_full_croissant(gid).loc[lambda df: df.dataset_name != 'evidence']\n",
    "    ])\n",
    "    .drop(['Example', 'field_join'], axis=1)\n",
    "    .sort_values(['dataset_name', 'field_id'])\n",
    "    .to_csv('annotated_new_evidence_columns.csv', index=False, sep=',')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "37c12ddb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1199"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pd.read_csv('annotated_new_evidence_columns.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d67d8b68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>field_id</th>\n",
       "      <th>column_name</th>\n",
       "      <th>field_join</th>\n",
       "      <th>column_description</th>\n",
       "      <th>foreign_key</th>\n",
       "      <th>bioregistry_prefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>evidence_cancer_biomarkers</td>\n",
       "      <td>evidence_cancer_biomarkers/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>evidence_cancer_gene_census</td>\n",
       "      <td>evidence_cancer_gene_census/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>evidence_cancer_gene_census</td>\n",
       "      <td>evidence_cancer_gene_census/directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>Effect of direction on target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>evidence_chembl</td>\n",
       "      <td>evidence_chembl/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>evidence_chembl</td>\n",
       "      <td>evidence_chembl/directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>Effect of direction on target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>evidence_clingen</td>\n",
       "      <td>evidence_clingen/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>evidence_crispr</td>\n",
       "      <td>evidence_crispr/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>evidence_crispr_screen</td>\n",
       "      <td>evidence_crispr_screen/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>evidence_encore</td>\n",
       "      <td>evidence_encore/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>evidence_europepmc</td>\n",
       "      <td>evidence_europepmc/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>evidence_eva</td>\n",
       "      <td>evidence_eva/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>evidence_eva</td>\n",
       "      <td>evidence_eva/directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>Effect of direction on target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>evidence_eva_somatic</td>\n",
       "      <td>evidence_eva_somatic/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>evidence_eva_somatic</td>\n",
       "      <td>evidence_eva_somatic/directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>Effect of direction on target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>evidence_expression_atlas</td>\n",
       "      <td>evidence_expression_atlas/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>evidence_gene_burden</td>\n",
       "      <td>evidence_gene_burden/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>evidence_gene_burden</td>\n",
       "      <td>evidence_gene_burden/directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>Effect of direction on target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>evidence_gene2phenotype</td>\n",
       "      <td>evidence_gene2phenotype/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>evidence_gene2phenotype</td>\n",
       "      <td>evidence_gene2phenotype/directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>Effect of direction on target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>evidence_genomics_england</td>\n",
       "      <td>evidence_genomics_england/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>evidence_gwas_credible_sets</td>\n",
       "      <td>evidence_gwas_credible_sets/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>evidence_impc</td>\n",
       "      <td>evidence_impc/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>evidence_impc</td>\n",
       "      <td>evidence_impc/directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>Effect of direction on target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>evidence_intogen</td>\n",
       "      <td>evidence_intogen/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413</th>\n",
       "      <td>evidence_intogen</td>\n",
       "      <td>evidence_intogen/directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>Effect of direction on target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>evidence_orphanet</td>\n",
       "      <td>evidence_orphanet/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>evidence_orphanet</td>\n",
       "      <td>evidence_orphanet/directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>directionOnTarget</td>\n",
       "      <td>Effect of direction on target</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>evidence_ot_crispr</td>\n",
       "      <td>evidence_ot_crispr/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482</th>\n",
       "      <td>evidence_ot_crispr_validation</td>\n",
       "      <td>evidence_ot_crispr_validation/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>evidence_reactome</td>\n",
       "      <td>evidence_reactome/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>evidence_uniprot_literature</td>\n",
       "      <td>evidence_uniprot_literature/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537</th>\n",
       "      <td>evidence_uniprot_variants</td>\n",
       "      <td>evidence_uniprot_variants/qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>qualityControls</td>\n",
       "      <td>Evidence quality flags</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      dataset_name  \\\n",
       "23      evidence_cancer_biomarkers   \n",
       "44     evidence_cancer_gene_census   \n",
       "50     evidence_cancer_gene_census   \n",
       "69                 evidence_chembl   \n",
       "75                 evidence_chembl   \n",
       "88                evidence_clingen   \n",
       "108                evidence_crispr   \n",
       "131         evidence_crispr_screen   \n",
       "164                evidence_encore   \n",
       "186             evidence_europepmc   \n",
       "213                   evidence_eva   \n",
       "219                   evidence_eva   \n",
       "241           evidence_eva_somatic   \n",
       "247           evidence_eva_somatic   \n",
       "262      evidence_expression_atlas   \n",
       "301           evidence_gene_burden   \n",
       "307           evidence_gene_burden   \n",
       "321        evidence_gene2phenotype   \n",
       "327        evidence_gene2phenotype   \n",
       "341      evidence_genomics_england   \n",
       "355    evidence_gwas_credible_sets   \n",
       "383                  evidence_impc   \n",
       "389                  evidence_impc   \n",
       "407               evidence_intogen   \n",
       "413               evidence_intogen   \n",
       "427              evidence_orphanet   \n",
       "433              evidence_orphanet   \n",
       "451             evidence_ot_crispr   \n",
       "482  evidence_ot_crispr_validation   \n",
       "503              evidence_reactome   \n",
       "519    evidence_uniprot_literature   \n",
       "537      evidence_uniprot_variants   \n",
       "\n",
       "                                          field_id        column_name  \\\n",
       "23      evidence_cancer_biomarkers/qualityControls    qualityControls   \n",
       "44     evidence_cancer_gene_census/qualityControls    qualityControls   \n",
       "50   evidence_cancer_gene_census/directionOnTarget  directionOnTarget   \n",
       "69                 evidence_chembl/qualityControls    qualityControls   \n",
       "75               evidence_chembl/directionOnTarget  directionOnTarget   \n",
       "88                evidence_clingen/qualityControls    qualityControls   \n",
       "108                evidence_crispr/qualityControls    qualityControls   \n",
       "131         evidence_crispr_screen/qualityControls    qualityControls   \n",
       "164                evidence_encore/qualityControls    qualityControls   \n",
       "186             evidence_europepmc/qualityControls    qualityControls   \n",
       "213                   evidence_eva/qualityControls    qualityControls   \n",
       "219                 evidence_eva/directionOnTarget  directionOnTarget   \n",
       "241           evidence_eva_somatic/qualityControls    qualityControls   \n",
       "247         evidence_eva_somatic/directionOnTarget  directionOnTarget   \n",
       "262      evidence_expression_atlas/qualityControls    qualityControls   \n",
       "301           evidence_gene_burden/qualityControls    qualityControls   \n",
       "307         evidence_gene_burden/directionOnTarget  directionOnTarget   \n",
       "321        evidence_gene2phenotype/qualityControls    qualityControls   \n",
       "327      evidence_gene2phenotype/directionOnTarget  directionOnTarget   \n",
       "341      evidence_genomics_england/qualityControls    qualityControls   \n",
       "355    evidence_gwas_credible_sets/qualityControls    qualityControls   \n",
       "383                  evidence_impc/qualityControls    qualityControls   \n",
       "389                evidence_impc/directionOnTarget  directionOnTarget   \n",
       "407               evidence_intogen/qualityControls    qualityControls   \n",
       "413             evidence_intogen/directionOnTarget  directionOnTarget   \n",
       "427              evidence_orphanet/qualityControls    qualityControls   \n",
       "433            evidence_orphanet/directionOnTarget  directionOnTarget   \n",
       "451             evidence_ot_crispr/qualityControls    qualityControls   \n",
       "482  evidence_ot_crispr_validation/qualityControls    qualityControls   \n",
       "503              evidence_reactome/qualityControls    qualityControls   \n",
       "519    evidence_uniprot_literature/qualityControls    qualityControls   \n",
       "537      evidence_uniprot_variants/qualityControls    qualityControls   \n",
       "\n",
       "            field_join             column_description foreign_key  \\\n",
       "23     qualityControls         Evidence quality flags         NaN   \n",
       "44     qualityControls         Evidence quality flags         NaN   \n",
       "50   directionOnTarget  Effect of direction on target         NaN   \n",
       "69     qualityControls         Evidence quality flags         NaN   \n",
       "75   directionOnTarget  Effect of direction on target         NaN   \n",
       "88     qualityControls         Evidence quality flags         NaN   \n",
       "108    qualityControls         Evidence quality flags         NaN   \n",
       "131    qualityControls         Evidence quality flags         NaN   \n",
       "164    qualityControls         Evidence quality flags         NaN   \n",
       "186    qualityControls         Evidence quality flags         NaN   \n",
       "213    qualityControls         Evidence quality flags         NaN   \n",
       "219  directionOnTarget  Effect of direction on target         NaN   \n",
       "241    qualityControls         Evidence quality flags         NaN   \n",
       "247  directionOnTarget  Effect of direction on target         NaN   \n",
       "262    qualityControls         Evidence quality flags         NaN   \n",
       "301    qualityControls         Evidence quality flags         NaN   \n",
       "307  directionOnTarget  Effect of direction on target         NaN   \n",
       "321    qualityControls         Evidence quality flags         NaN   \n",
       "327  directionOnTarget  Effect of direction on target         NaN   \n",
       "341    qualityControls         Evidence quality flags         NaN   \n",
       "355    qualityControls         Evidence quality flags         NaN   \n",
       "383    qualityControls         Evidence quality flags         NaN   \n",
       "389  directionOnTarget  Effect of direction on target         NaN   \n",
       "407    qualityControls         Evidence quality flags         NaN   \n",
       "413  directionOnTarget  Effect of direction on target         NaN   \n",
       "427    qualityControls         Evidence quality flags         NaN   \n",
       "433  directionOnTarget  Effect of direction on target         NaN   \n",
       "451    qualityControls         Evidence quality flags         NaN   \n",
       "482    qualityControls         Evidence quality flags         NaN   \n",
       "503    qualityControls         Evidence quality flags         NaN   \n",
       "519    qualityControls         Evidence quality flags         NaN   \n",
       "537    qualityControls         Evidence quality flags         NaN   \n",
       "\n",
       "    bioregistry_prefix  \n",
       "23                 NaN  \n",
       "44                 NaN  \n",
       "50                 NaN  \n",
       "69                 NaN  \n",
       "75                 NaN  \n",
       "88                 NaN  \n",
       "108                NaN  \n",
       "131                NaN  \n",
       "164                NaN  \n",
       "186                NaN  \n",
       "213                NaN  \n",
       "219                NaN  \n",
       "241                NaN  \n",
       "247                NaN  \n",
       "262                NaN  \n",
       "301                NaN  \n",
       "307                NaN  \n",
       "321                NaN  \n",
       "327                NaN  \n",
       "341                NaN  \n",
       "355                NaN  \n",
       "383                NaN  \n",
       "389                NaN  \n",
       "407                NaN  \n",
       "413                NaN  \n",
       "427                NaN  \n",
       "433                NaN  \n",
       "451                NaN  \n",
       "482                NaN  \n",
       "503                NaN  \n",
       "519                NaN  \n",
       "537                NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(\n",
    "    updated_evidence_table\n",
    "    .assign(\n",
    "        column_description = lambda df: df.apply(\n",
    "            lambda row: \n",
    "              \"Evidence quality flags\" if row['column_name'] == 'qualityControls' else \n",
    "              \"Effect of direction on target\" if row['column_name'] == 'directionOnTarget' else\n",
    "              row['column_description'],\n",
    "            axis=1\n",
    "        )\n",
    "    )\n",
    "    .loc[updated_evidence_table.column_description.isna()]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93b23df9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5bf1d5a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ot-croissant",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
