{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "25609461",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         datasourceId                    source configuration  \\\n",
      "0   cancer_biomarkers  cancerbiomarkers.json.gz            OK   \n",
      "1  cancer_gene_census            cosmic.json.gz           OK    \n",
      "2              chembl            chembl.json.gz            OK   \n",
      "3             clingen           clingen.json.gz            OK   \n",
      "4              crispr            crispr.json.gz            OK   \n",
      "\n",
      "  Direction of effect  run time (s) isScoreOK    valid old valid invalid  \\\n",
      "0                 NaN          47.0        OK    1,300     1,300      49   \n",
      "1                  OK          55.0        OK   91,572    91,572  11,300   \n",
      "2                  OK          27.0        OK  573,103   573,103  18,243   \n",
      "3                 NaN          18.0        OK    3,047     3,047     292   \n",
      "4                 NaN          52.0        OK      517       517       0   \n",
      "\n",
      "  old invalid comment                       output  \\\n",
      "0          49     NaN   evidence_cancer_biomarkers   \n",
      "1      11,300     NaN  evidence_cancer_gene_census   \n",
      "2      18,243     NaN              evidence_chembl   \n",
      "3         292     NaN             evidence_clingen   \n",
      "4           0     NaN              evidence_crispr   \n",
      "\n",
      "                      niceName  \\\n",
      "0    Cancer biomarker evidence   \n",
      "1  Cancer Gene Census evidence   \n",
      "2              ChEMBL evidence   \n",
      "3             ClinGen evidence   \n",
      "4        ProjectScore evidence   \n",
      "\n",
      "                                         description  \n",
      "0  Disease-target evidence based on cancer biomar...  \n",
      "1  Disease-target evidence based on Cancer Gene C...  \n",
      "2  Disease-target evidence based on drug indicati...  \n",
      "3  Disease-target evidence from Gene–Disease Vali...  \n",
      "4  Diease-target evidence from unified analysis o...  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'cancer_biomarkers': {'output': 'evidence_cancer_biomarkers',\n",
       "  'niceName': 'Cancer biomarker evidence',\n",
       "  'description': 'Disease-target evidence based on cancer biomarkers from the Cancer Genome Interpreter'},\n",
       " 'cancer_gene_census': {'output': 'evidence_cancer_gene_census',\n",
       "  'niceName': 'Cancer Gene Census evidence',\n",
       "  'description': 'Disease-target evidence based on Cancer Gene Census provided by COSMIC'},\n",
       " 'chembl': {'output': 'evidence_chembl',\n",
       "  'niceName': 'ChEMBL evidence',\n",
       "  'description': 'Disease-target evidence based on drug indications, clinical trials and mechanism of action provided by ChEMBL'},\n",
       " 'clingen': {'output': 'evidence_clingen',\n",
       "  'niceName': 'ClinGen evidence',\n",
       "  'description': 'Disease-target evidence from Gene–Disease Validity Curation of ClinGen'},\n",
       " 'crispr': {'output': 'evidence_crispr',\n",
       "  'niceName': 'ProjectScore evidence',\n",
       "  'description': 'Diease-target evidence from unified analysis of whole-genome CRISPR/Cas9 screens from ProjectScore project'},\n",
       " 'crispr_screen': {'output': 'evidence_crispr_screen',\n",
       "  'niceName': 'CRISP screen evidence',\n",
       "  'description': 'Disease-target evidence derived from systematic gene perturation assays of CRISPR knock-out screens'},\n",
       " 'encore': {'output': 'evidence_encore',\n",
       "  'niceName': 'Open Targets - Encore evidence',\n",
       "  'description': 'Disease-target evidence based on somatic variation from ClinVar, provided by the European Variation Archive'},\n",
       " 'europepmc': {'output': 'evidence_europepmc',\n",
       "  'niceName': 'EuropePMC evidence',\n",
       "  'description': 'Disease-target evidence derived from literature mining provided by Europe PMC'},\n",
       " 'eva': {'output': 'evidence_eva',\n",
       "  'niceName': 'ClinVar germ-line evidence',\n",
       "  'description': 'Disease-target evidence based on germline variation from ClinVar, provided by the European Variation Archive'},\n",
       " 'eva_somatic': {'output': 'evidence_eva_somatic',\n",
       "  'niceName': 'ClinVar somatic evidence',\n",
       "  'description': 'Disease-target evidence based on somatic variation from ClinVar, provided by the European Variation Archive'},\n",
       " 'expression_atlas': {'output': 'evidence_expression_atlas',\n",
       "  'niceName': 'Differential expression evidence',\n",
       "  'description': 'Disease-target evidence based on differential expression data provided by Expression Atlas'},\n",
       " 'gene_burden': {'output': 'evidence_gene_burden',\n",
       "  'niceName': 'Gene burden evidence',\n",
       "  'description': 'Disease-target evidence based on testing aggregated effect of variants, curated from the literature by the Open Targets team'},\n",
       " 'gene2phenotype': {'output': 'evidence_gene2phenotype',\n",
       "  'niceName': 'Gene2Phenotype evidence',\n",
       "  'description': 'Disease-target evidence curated by the Gene2Phenotype consortium'},\n",
       " 'genomics_england': {'output': 'evidence_genomics_england',\n",
       "  'niceName': 'Genomics England - PanelApp evidence',\n",
       "  'description': 'Curated disease-target evidence provided by Genomics England PanelApp'},\n",
       " 'gwas_credible_set': {'output': 'evidence_gwas_credible_sets',\n",
       "  'niceName': 'GWAS credible-set evidence',\n",
       "  'description': 'Disease-target evidence based on Locus-to-Gene prediction on GWAS credible sets'},\n",
       " 'impc': {'output': 'evidence_impc',\n",
       "  'niceName': 'IMPC evidence',\n",
       "  'description': 'Disease-target evidence based on mouse knock-out models, where mouse phenotypes are mapped to human diseases by PhenoDigm algorithm. Data provided by IMPC'},\n",
       " 'intogen': {'output': 'evidence_intogen',\n",
       "  'niceName': 'IntOGen evidence',\n",
       "  'description': 'Cancer-driver gene evidence based on the analysis of large-scale mutational data from IntOGen'},\n",
       " 'orphanet': {'output': 'evidence_orphanet',\n",
       "  'niceName': 'Orphanet evidence',\n",
       "  'description': 'Curated collection of rare disase-target evidence from Orphanet'},\n",
       " 'ot_crispr': {'output': 'evidence_ot_crispr',\n",
       "  'niceName': 'Open Targets - CRISPR screen evidence',\n",
       "  'description': 'Pre-publication CRISPR/Cas9 perturation screens provided by Open Targets experimental projects'},\n",
       " 'ot_crispr_validation': {'output': 'evidence_ot_crispr_validation',\n",
       "  'niceName': 'Open Targets - Validation Lab evidence',\n",
       "  'description': 'Disease-target evidence derived from CRISPR/Cas9 perturation screens provided by Open Targets Validation lab'},\n",
       " 'reactome': {'output': 'evidence_reactome',\n",
       "  'niceName': 'Reactome evidence',\n",
       "  'description': 'Manually curated disease-target evidence based on reaction pathways sourced from Reactome'},\n",
       " 'uniprot_literature': {'output': 'evidence_uniprot_literature',\n",
       "  'niceName': 'Uniprot literature evidence',\n",
       "  'description': 'Protein-disease evidence based on literature curated by the Uniprot team'},\n",
       " 'uniprot_variants': {'output': 'evidence_uniprot_variants',\n",
       "  'niceName': 'Uniprot variants evidence',\n",
       "  'description': 'Disease-target evidence based on disease causing protein sequence alteration curated by the Uniprot team'}}"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from typing import Any\n",
    "import json \n",
    "\n",
    "# Extract table from spreadsheet:\n",
    "df = (\n",
    "    pd.read_csv('https://docs.google.com/spreadsheets/d/1G3zq-aJg3uKBz3VZUvGV8yYVrECtpQ9SJoGX6X3B-ms/export?format=csv', sep=',')\n",
    "    .loc[lambda df: df.output.notna()]\n",
    ")\n",
    "print(df.head())\n",
    "\n",
    "# Convert table to mapping:\n",
    "datasoruce_mapping = (\n",
    "    df[['output', 'niceName', 'description']]\n",
    "    .set_index(\n",
    "        pd.Series(df.datasourceId.to_list())\n",
    "    )\n",
    "    .to_dict(orient='index')\n",
    ")\n",
    "\n",
    "datasoruce_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "ee0986fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "  {\n",
      "    \"id\": \"evidence_cancer_biomarkers\",\n",
      "    \"nice_name\": \"Cancer biomarker evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_cancer_biomarkers/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence based on cancer biomarkers from the Cancer Genome Interpreter\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_cancer_gene_census\",\n",
      "    \"nice_name\": \"Cancer Gene Census evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_cancer_gene_census/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence based on Cancer Gene Census provided by COSMIC\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_chembl\",\n",
      "    \"nice_name\": \"ChEMBL evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_chembl/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence based on drug indications, clinical trials and mechanism of action provided by ChEMBL\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_clingen\",\n",
      "    \"nice_name\": \"ClinGen evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_clingen/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence from Gene\\u2013Disease Validity Curation of ClinGen\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_crispr\",\n",
      "    \"nice_name\": \"ProjectScore evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_crispr/id\"\n",
      "    ],\n",
      "    \"description\": \"Diease-target evidence from unified analysis of whole-genome CRISPR/Cas9 screens from ProjectScore project\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_crispr_screen\",\n",
      "    \"nice_name\": \"CRISP screen evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_crispr_screen/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence derived from systematic gene perturation assays of CRISPR knock-out screens\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_encore\",\n",
      "    \"nice_name\": \"Open Targets - Encore evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_encore/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence based on somatic variation from ClinVar, provided by the European Variation Archive\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_europepmc\",\n",
      "    \"nice_name\": \"EuropePMC evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_europepmc/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence derived from literature mining provided by Europe PMC\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_eva\",\n",
      "    \"nice_name\": \"ClinVar germ-line evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_eva/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence based on germline variation from ClinVar, provided by the European Variation Archive\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_eva_somatic\",\n",
      "    \"nice_name\": \"ClinVar somatic evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_eva_somatic/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence based on somatic variation from ClinVar, provided by the European Variation Archive\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_expression_atlas\",\n",
      "    \"nice_name\": \"Differential expression evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_expression_atlas/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence based on differential expression data provided by Expression Atlas\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_gene_burden\",\n",
      "    \"nice_name\": \"Gene burden evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_gene_burden/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence based on testing aggregated effect of variants, curated from the literature by the Open Targets team\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_gene2phenotype\",\n",
      "    \"nice_name\": \"Gene2Phenotype evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_gene2phenotype/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence curated by the Gene2Phenotype consortium\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_genomics_england\",\n",
      "    \"nice_name\": \"Genomics England - PanelApp evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_genomics_england/id\"\n",
      "    ],\n",
      "    \"description\": \"Curated disease-target evidence provided by Genomics England PanelApp\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_gwas_credible_sets\",\n",
      "    \"nice_name\": \"GWAS credible-set evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_gwas_credible_sets/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence based on Locus-to-Gene prediction on GWAS credible sets\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_impc\",\n",
      "    \"nice_name\": \"IMPC evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_impc/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence based on mouse knock-out models, where mouse phenotypes are mapped to human diseases by PhenoDigm algorithm. Data provided by IMPC\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_intogen\",\n",
      "    \"nice_name\": \"IntOGen evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_intogen/id\"\n",
      "    ],\n",
      "    \"description\": \"Cancer-driver gene evidence based on the analysis of large-scale mutational data from IntOGen\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_orphanet\",\n",
      "    \"nice_name\": \"Orphanet evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_orphanet/id\"\n",
      "    ],\n",
      "    \"description\": \"Curated collection of rare disase-target evidence from Orphanet\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_ot_crispr\",\n",
      "    \"nice_name\": \"Open Targets - CRISPR screen evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_ot_crispr/id\"\n",
      "    ],\n",
      "    \"description\": \"Pre-publication CRISPR/Cas9 perturation screens provided by Open Targets experimental projects\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_ot_crispr_validation\",\n",
      "    \"nice_name\": \"Open Targets - Validation Lab evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_ot_crispr_validation/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence derived from CRISPR/Cas9 perturation screens provided by Open Targets Validation lab\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_reactome\",\n",
      "    \"nice_name\": \"Reactome evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_reactome/id\"\n",
      "    ],\n",
      "    \"description\": \"Manually curated disease-target evidence based on reaction pathways sourced from Reactome\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_uniprot_literature\",\n",
      "    \"nice_name\": \"Uniprot literature evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_uniprot_literature/id\"\n",
      "    ],\n",
      "    \"description\": \"Protein-disease evidence based on literature curated by the Uniprot team\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  },\n",
      "  {\n",
      "    \"id\": \"evidence_uniprot_variants\",\n",
      "    \"nice_name\": \"Uniprot variants evidence\",\n",
      "    \"key\": [\n",
      "      \"evidence_uniprot_variants/id\"\n",
      "    ],\n",
      "    \"description\": \"Disease-target evidence based on disease causing protein sequence alteration curated by the Uniprot team\",\n",
      "    \"tags\": [\n",
      "      \"Target-Disease\"\n",
      "    ]\n",
      "  }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "def process_dataset(output: str, nice_name:str, description:str) -> dict[str,Any]:\n",
    "    return {\n",
    "        \"id\": output,\n",
    "        \"nice_name\": nice_name,\n",
    "        \"key\": [ f\"{output}/id\"],\n",
    "        \"description\": description,\n",
    "        \"tags\": [\"Target-Disease\"]\n",
    "    }\n",
    "\n",
    "print(\n",
    "    json.dumps(\n",
    "        [\n",
    "            process_dataset(value['output'], value['niceName'], value['description'])\n",
    "            for value in datasoruce_mapping.values()\n",
    "        ],\n",
    "        indent=2\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ad7d7c6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ot-croissant",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
